---
title: "28.BLUP.GWAS"
author: "Daniele Filiault"
date: "2/10/2020"
output: html_document
---

```{r setup, include=FALSE}
library(knitr)
knitr::opts_chunk$set(echo = TRUE)
read_chunk("./scripts/998.Ben_scripts/GWA_functions.R")

library(dplyr)
library(tidyr)
library(lattice)

```

## Running GWAS for BLUPs calculated in 27.fitness.BLUPS.Rmd
## Uses gemma and the setup thereof in markdown 01.CG.DLF.initial.Rmd


### 1a. Make phenotype file for blups
```{r make pheno files for GWAS}
### read in blups
b.dat <- read.csv("./data/marginal.blups.csv")
b.dat <- b.dat %>% pivot_wider(id_cols=id,names_from=c(exp,year),values_from=fb)%>%as.data.frame()

##make a phenotype file
pref="./GWA/snps/sweden_200_MAF10.bimbam"
dat <- b.dat
row.names(dat)=paste(dat$id)
dat=dat[,-1]
acclist=scan(paste(pref, ".acclist", sep=""), sep=",")

##make a table of phenos in the same order as the genotype
phen=dat[paste(acclist), ]
colnames(phen) <- paste("blup", colnames(phen), sep="_")
phenfile=paste(pref, ".blups.marginal.phen.txt", sep="")

#write.table(phen, phenfile, sep="\t", row.names=F, col.names=F)
```

### 1b. Run marginal GWAS with gemma
```{r eval=FALSE, results='hide'}

##run gemma using gemma-wrapper
## to get system() to work in .Rstudio, you need to open .RStudio in the commmand line to inhert setup.  
## command is:
## open -a RStudio

#system("mkdir ./GWA/output_gemma_marginal/")
#system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache -- -gk -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " > ./GWA/output_gemma_marginal/K.json", sep=""))

#Show in New WindowClear OutputExpand/Collapse Output
#Read 200 items
#Show in New WindowClear OutputExpand/Collapse Output
##Read 10 items
#Show in New WindowClear OutputExpand/Collapse Output
#Read 10 items

for(n in 1:ncol(phen)){
    pheno=colnames(phen)[n]
    print(pheno)
    system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_marginal/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", n , " > ./GWA/output_gemma_marginal/out_marginal_", pheno, sep=""))
}
```


### 1c. Make manhattan plots and output 

```{r eval=FALSE, include=TRUE}
source("./scripts/998.Ben_scripts/GWA_functions.R")
ph2=data.frame(phen=colnames(phen), ph2=NA) # will record heritabilites
gwsignif=list()
for(n in 1:ncol(phen)){
    pheno=colnames(phen)[n]
    print(pheno)
    x=unlist(strsplit(scan(paste("./GWA/output_gemma_marginal/out_marginal_", pheno,sep=""), what="character")[10], ","))[7]  ### had to play with this a bit to deal with a warning message that made my output different from Ben's
    f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
    f2=gsub("assoc", "log", f)
    ph2$ph2[n]=ph2gemma(f2)
    p=read.delim(f, header=T, sep="\t")
    save(p, file=paste("./res/gemma_marginal/gemma_lmm_", pheno,".rds", sep="")) #edited 28SEPT20
    ##make a manhattan plot
    colnames(p)[3]="pos"
    colnames(p)=gsub("p_lrt", "pval", colnames(p))
    p$score=-log10(p$pval)
    p=p[p$af>=0.1,]
    p$fdr_pval=p.adjust(p$pval, "fdr")
    gwsignif[[pheno]]=p[p$fdr_pval<=0.05,]
    jpeg(paste("./GWA/manhattans_marginal/lmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
    manhattan(p)
    dev.off()
    if(n==1){
        beta=p[, c("rs","beta")]
        se=p[, c("rs", "se")]
    }else{
        beta=merge(p[, c("rs","beta")], beta, by="rs", all=T)
        se=merge(p[, c("rs","se")], se, by="rs", all=T)
    }
}

#output heritabilities
write.csv(ph2, file="./data/blup.fitness.gemma.H2.csv",quote=FALSE,row.names=FALSE)
```


## 2.  Mvlmm for 2011 for all 4 sites (4 phenos total)

Run GWAS
```{r, 2011gwas, eval=FALSE}
n=c(1,3,5,7) ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```

Make manhattan plot and extract pvals and betas
```{r 2011 manhattan}
##make a manhattan plot
n=c(1,3,5,7)
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()

#fig_nums("manhattan1", caption="Manhattan plot - all sites fitness 2011")
#include_graphics(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""))
```

Output betas for use elsewhere
```{r output betas 2011, eval=FALSE}
blup.betas2011 <- p
save(blup.betas2011, file="./data/blup.betas2011.Rdat")
rm(blup.betas2011)
```

## 3.  Mvlmm for 2012 for all 4 sites (4 phenos total)

Run GWAS
```{r, 2012gwas, eval=FALSE}
n=c(2,4,6,8)  ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```


Make manhattan plot and extract pvals and betas
```{r 2012 manhattan}
##make a manhattan plot
n=c(2,4,6,8) 
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()

#fig_nums("manhattan2", caption="Manhattan plot - all sites fitness 2012")
#include_graphics(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""))
```

Output betas for use elsewhere
```{r output betas 2012, eval=FALSE}
blup.betas2012 <- p
save(blup.betas2012, file="./data/blup.betas2012.Rdat")
#rm(betas2012)
```


## 4. mvlmm ADA blups, both years
```{r, ADA_gwas, eval=FALSE}
n=c(1,2)  ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```

Make manhattan plot and extract pvals and betas
```{r ADA manhattan}
##make a manhattan plot
n=c(1,2) 
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()
```

Output betas for use elsewhere
```{r output betas ADA, eval=FALSE}
blup.betasADA <- p
save(blup.betasADA, file="./data/blup.betasADA.Rdat")
#rm(betasADA)
```



## 5. mvlmm RAM blups, both years
```{r, RAM_gwas, eval=FALSE}
n=c(3,4)  ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```

Make manhattan plot and extract pvals and betas
```{r RAM manhattan}
##make a manhattan plot
n=c(3,4) 
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()
```

Output betas for use elsewhere
```{r output betas RAM, eval=FALSE}
blup.betasRAM <- p
save(blup.betasRAM, file="./data/blup.betasRAM.Rdat")
#rm(betasRAM)
```

## 6. mvlmm ULL blups, both years
```{r, ULL_gwas, eval=FALSE}
n=c(7,8)  ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```

Make manhattan plot and extract pvals and betas
```{r ULL manhattan}
##make a manhattan plot
n=c(7,8) 
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()
```

Output betas for use elsewhere
```{r output betas ULL, eval=FALSE}
blup.betasULL <- p
save(blup.betasULL, file="./data/blup.betasULL.Rdat")
#rm(betasULL)
```

## 7. mvlmm RAT blups, both years
```{r, RAT_gwas, eval=FALSE}
n=c(5,6)  ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```

Make manhattan plot and extract pvals and betas
```{r RAT manhattan}
##make a manhattan plot
n=c(5,6) 
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()
```

Output betas for use elsewhere
```{r output betas RAT, eval=FALSE}
blup.betasRAT <- p
save(blup.betasRAT, file="./data/blup.betasRAT.Rdat")
#rm(betasADA)
```


## 8. mvlmm all north experiments
```{r, north_gwas}
n=c(1:4)  ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```


Make manhattan plot and extract pvals and betas
```{r north manhattan}
##make a manhattan plot
n=c(1:4) 
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()
```

Output betas for use elsewhere
```{r output betas north}
blup.betasnorth <- p
save(blup.betasnorth, file="./data/blup.betasnorth.Rdat")
```

## 9. mvlmm all south experiments
```{r, south_gwas}
n=c(5:8)  ## column numbers of phen file to include
sel=paste(n, collapse=" ")
pheno=paste(colnames(phen)[n], collapse="_")
system(paste("/Users/daniele.filiault/.gem/ruby/2.0.0/bio-gemma-wrapper-0.98.1/bin/gemma-wrapper --loco --json --cache-dir ./GWA/.gemma-cache --input ./GWA/output_gemma_loco/K.json -- -lmm 4 -g ", pref, ".geno.gz -a ", pref, ".map -p ", phenfile, " -n ", sel , " > ./GWA/output_gemma_mvlmm/out_mvlmm_", pheno, sep=""))

```
Make manhattan plot and extract pvals and betas
```{r south manhattan}
##make a manhattan plot
n=c(5:8) 
pheno=paste(colnames(phen)[n], collapse="_")
x=unlist(strsplit(scan(paste("./GWA/output_gemma_mvlmm/out_mvlmm_", pheno,sep=""), what="character")[10], ","))[7]
### had to play with this a bit to deal with a warning message that made my output different from Ben'sf=substring(gsub("]]", "", x), 2, (nchar(x)-3))
### ulimate output is the corresponding assoc.txt file in the gemma-cache
f=substring(gsub("]]", "", x), 2, (nchar(x)-3))
p=read.delim(f, header=T, sep="\t")
##make a manhattan plot
colnames(p)[3]="pos"
colnames(p)=gsub("p_wald", "pval", colnames(p))
p$score=-log10(p$pval)
p=p[p$af>=0.1,]
p$fdr_pval=p.adjust(p$pval, "fdr")
mlmmsignif=p[p$fdr_pval<=0.1,]
source("./scripts/GWA_functions.R")
jpeg(paste("./GWA/manhattans/mvlmm_gwa_", pheno, ".jpeg", sep=""),res=600, unit="in", width=8, height=4, pointsize=12)
manhattan(p)
dev.off()
```

Output betas for use elsewhere
```{r output betas north}
blup.betassouth <- p
save(blup.betassouth, file="./data/blup.betassouth.Rdat")
```

## 10.  Make QQplots of by year data
```{r qqplots, eval=FALSE}
load(file="./data/blup.betas2012.Rdat")
load(file="./data/blup.betas2011.Rdat")

jpeg("./figures/qqplot.fitness2011gemma.jpg",res=600, unit="in", width=5, height=5, pointsize=12)
qqmath(-log10(blup.betas2011$pval), distribution=function(x){-log10(qunif(1-x))}, main="QQ plot BLUPs 2011",panel = function(x, ...) { 
    panel.qqmathline(x, ...) 
    panel.qqmath(x, ...) 
  })
dev.off()

jpeg("./figures/qqplot.fitness2012gemma.jpg",res=600, unit="in", width=5, height=5, pointsize=12)
qqmath(-log10(blup.betas2012$pval), distribution=function(x){-log10(qunif(1-x))}, main="QQ plot BLUPs 2012",panel = function(x, ...) { 
    panel.qqmathline(x, ...) 
    panel.qqmath(x, ...) 
  })
dev.off()
```

## 10. Filter GWAS by pvalue for overlap with selection scans
Need to decide on some cutoff criteria.  
Generate with -log10pval 4 and 5
```{r filter year blups }
filt.pval <- function(upname, cutoff){
  upvar <- get(upname)
  outvar <- upvar[upvar$score>=cutoff,c(1,3)]
  outfile <- paste("./data/",upname,".",cutoff,".pval.cutoff.csv",sep="")
  print(outfile)
  write.csv(outvar, file=outfile,row.names=FALSE)
}

a <- filt.pval(upname="blup.betas2011", cutoff=4)
a <- filt.pval(upname="blup.betas2011", cutoff=5)
a <- filt.pval(upname="blup.betas2012", cutoff=4)
a <-filt.pval(upname="blup.betas2012", cutoff=5)
```


