---
title: "41.gene.annot.in.windows"
author: "Daniele Filiault"
date: "7/28/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

setwd("/Volumes/field_experiments/adaptation_sweden/common.gardens/")
```

## Introduction

Devloping a script to read a csv of chr + pos values and get genes and annotation in a +/- window around all these snps.

### Set variables
```{r set variables}
## file with chr and pos as first two fields
snps.file <- "/Volumes/field_experiments/adaptation_sweden/common.gardens/data/003.selection.scans/001.formatted.for.limix/christian.intervals.csv"
# TAIR10 gff3 file, already parsed to only "genes"
gff.file <- "/Volumes/field_experiments/001.common.reference.files/001.TAIR10.genome/genes.txt"
# TAIR aliases file (gene names)
gene.names.file <-  "/Volumes/field_experiments/001.common.reference.files/001.TAIR10.genome/gene_aliases_20190630.txt"
# GO slim terms
go.slim.file <- "/Volumes/field_experiments/001.common.reference.files/001.TAIR10.genome/ATH_GO_GOSLIM.txt"
# Araport 11 annotation
ara.file <- "/Volumes/field_experiments/001.common.reference.files/001.TAIR10.genome/Araport11_functional_descriptions_20190630.txt"
# window size
win.size = 20000
# making a separate folder to keep output
out.dir="/Volumes/field_experiments/adaptation_sweden/common.gardens/data/005.gene.list.annotations/"
```

### read in and format data
```{r read data}
snps <- read.csv(snps.file)

gff <- read.table(gff.file, stringsAsFactors=FALSE)
atg <- unlist(lapply(strsplit(gff[,9],"="),function(x){x[4]}))
gff <- gff[,c(1,4,5)]
gff <- cbind(gff, atg)
gff$atg <- as.character(gff$atg)
colnames(gff) <- c("chr", "start","stop","name")
gff$chr <- gsub("Chr","",gff$chr)

gene.names <- read.table(gene.names.file, stringsAsFactors=FALSE,fill=TRUE,row.names=NULL, sep="\t",quote="", header=TRUE)
# name, symbol, full_name

go.slim <- read.table(go.slim.file, stringsAsFactors=FALSE, sep="\t",fill=TRUE, quote="")
go.slim <- go.slim[grep("AT", go.slim[,1]),]
go.slim <- go.slim[sapply(go.slim[,1],nchar)==9,] ### subsets to proper atg names
colnames(go.slim)[1] <- "name"
go.slim <- go.slim[,c(1,4,5,9)]

## the formatting of this ara file is a hot mess, not properly formatted.  However, I will just use it as is since this is only to get an idea of whether there are candidate genes in a window or not
ara <- read.table(ara.file, stringsAsFactors=FALSE, sep="\t", fill=TRUE, header=TRUE)
ara$name <- unlist(lapply(strsplit(ara[,1],".", fixed=TRUE),function(x){x[1]}))
ara <- ara[sapply(ara[,1],nchar)==9,] ### subsets to proper atg names.  there is a lot of weirdly-formatted crap here, but ignoring it because just using this to get an overview and do some cherry-picking
```

### develop script to get info for each snp in the snps file
```{r script development}
get.annot <- function(snps, window.size){
  out.dat <- as.list(1:nrow(snps))

  for(up in 1:nrow(snps)){
    up.dat <- snps[up,]
    #first step, use gff to get gene names of genes in window around this snps
    sstart <- up.dat$pos-win.size
    sstop <- up.dat$pos+win.size
    schr <- up.dat$chr
    up.bp <- sstart:sstop
    up.gs <- gff[gff$chr==schr,] 
    up.gs <- up.gs[up.gs$start%in%up.bp | up.gs$stop%in%up.bp,]
    # now, loop through each gene and get info from all three sources
    allg.info <- as.list(1:nrow(up.gs))
    for(up.g in 1:nrow(up.gs)){
      up.name <- up.gs[up.g,4]
      up.gn <- gene.names[gene.names$name==up.name,]
      up.go <- go.slim[go.slim$name==up.name,]
      up.go <- unique(up.go[,3])
      up.ara <- ara[ara$name==up.name,]
      out.g <- as.list(rep(NA,3))
      out.g[[1]] <- up.gn
      out.g[[2]] <- up.go
      out.g[[3]] <- up.ara
      allg.info[[up.g]] <- out.g
    }
    names(allg.info) <- up.gs$name
  out.dat[[up]] <- allg.info 
  }
  names(out.dat) <- paste(snps$chr, snps$pos, sep="_")
  return(out.dat)
}
```

```{r test script}

test.annot <- get.annot(snps=snps, window.size=window.size)

gene.n <- lapply(test.annot, function(x){lapply(x, function(y){y[[1]]})})
gene.list <- lapply(gene.n, function(x){do.call(rbind, x)})
### this extracts the gene names and symbols of genes that have them...
### it's a bit easier to quickly parse for specific things.
```

So this script works!  Now should go through GWAS for each experiment and just look at top hits to make sure not hitting something super interesting...
Easiest is to do this inline in this markdown, so can keep potentially interesting hits in one place.  
Just remember to save .Rdata output for these guys afterwards???  
Also could do a script to look for GO enrichment, but I doubt would see anything.

### get pvalues for all marginal tests
```{r load pvalue data}

gwas.res.files <- c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds","./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")

load(gwas.res.files[1])

pvals <- p[,c(1,3)]
betas <- pvals

for(up in 1:length(gwas.res.files)){
  up.gwa.file <- gwas.res.files[up]
  up.exp <- gsub("./res.gemma_marginal/gemma_lmm_blup_","",up.gwa.file)
  up.exp <- gsub(".rds","",up.exp)
  load(gwas.res.files[up])
  out.p <- p[,c(1,3,13)]
  colnames(out.p)[3] <- up.exp
  pvals <- merge(pvals,out.p)
  out.b <- p[,c(1,3,8)]
  colnames(out.b)[3] <- up.exp
  betas <- merge(betas,out.b)
}
colnames(pvals)[2] <- "pos"
colnames(betas)[2] <- "pos"

```

### for each experiment, take pvalue cutoff, generate gene lists

```{r get annot per exp, sig 5, window20000}
sig.cutoff <- 5
window.size <- 20000

for(up in 3:10){
  print(up)
  up.exp <- colnames(pvals)[up]
  print(up.exp)
  up.snps <- pvals[-log10(pvals[,up])>=sig.cutoff,]
  if(nrow(up.snps)>0){
    up.annot <- get.annot(snps=up.snps[,1:2], window.size=window.size)
    gene.n <- lapply(up.annot, function(x){lapply(x, function(y){y[[1]]})})
    up.gene.list <- lapply(gene.n, function(x){do.call(rbind, x)})
    for(up.l in 1:length(up.gene.list)){
      if(nrow(up.gene.list[[up.l]])==0){up.gene.list[[up.l]][1,] <- rep(NA,3)}
      up.gene.list[[up.l]] <- cbind(up.gene.list[[up.l]], names(up.gene.list)[up.l])
      }
    up.gene.list <- do.call(rbind,up.gene.list)
    colnames(up.gene.list)[4] <- "snp"
    rownames(up.gene.list) <- 1:nrow(up.gene.list)
    up.gene.list <- up.gene.list[order(up.gene.list$name),]
    up.gene.list$exp <- up.exp
    annot.file <- paste(out.dir,up.exp,".sig",sig.cutoff,".win",window.size,".annot.Rdata",sep="")
    save(up.annot, file=annot.file)
    gl.file <- paste(out.dir,up.exp,".sig",sig.cutoff,".win",window.size,".genes.tsv",sep="")
    write.table(up.gene.list, file=gl.file, quote=FALSE, row.names=FALSE, sep="\t")
  }
}
```

```{r get annot per exp, sig 5, window20000}
sig.cutoff <- 4
window.size <- 20000

for(up in 3:10){
  print(up)
  up.exp <- colnames(pvals)[up]
  print(up.exp)
  up.snps <- pvals[-log10(pvals[,up])>=sig.cutoff,]
  if(nrow(up.snps)>0){
    up.annot <- get.annot(snps=up.snps[,1:2], window.size=window.size)
    gene.n <- lapply(up.annot, function(x){lapply(x, function(y){y[[1]]})})
    up.gene.list <- lapply(gene.n, function(x){do.call(rbind, x)})
    for(up.l in 1:length(up.gene.list)){
      if(nrow(up.gene.list[[up.l]])==0){up.gene.list[[up.l]][1,] <- rep(NA,3)}
      up.gene.list[[up.l]] <- cbind(up.gene.list[[up.l]], names(up.gene.list)[up.l])
      }
    up.gene.list <- do.call(rbind,up.gene.list)
    colnames(up.gene.list)[4] <- "snp"
    rownames(up.gene.list) <- 1:nrow(up.gene.list)
    up.gene.list <- up.gene.list[order(up.gene.list$name),]
    up.gene.list$exp <- up.exp
    annot.file <- paste(out.dir,up.exp,".sig",sig.cutoff,".win",window.size,".annot.Rdata",sep="")
    save(up.annot, file=annot.file)
    gl.file <- paste(out.dir,up.exp,".sig",sig.cutoff,".win",window.size,".genes.tsv",sep="")
    write.table(up.gene.list, file=gl.file, quote=FALSE, row.names=FALSE, sep="\t")
  }
}
```

## Now this is a treasure hunt operation.

