---
title: "47.fitnessGWAS.by.AFD"
author: "DLF"
date: "9/22/2020 and 01/26/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
library(gridExtra)
library(ggpmisc)
library(tidyr)
library(ggpubr)
library(cowplot)
library(dplyr)
library(viridis)
library(rcompanion)
library(multcompView)
library(spdep)
source("99.geo.helper.fxns.R") ### has my genome plot function.

#setwd("/groups/nordborg/projects/field_experiments/adaptation_sweden/common.gardens")
setwd("/Volumes/field_experiments/adaptation_sweden/common.gardens/")
```

## Introduction

Look at relationship between AFD between K groups and p-values/betas from fitness GWAS run in gemma  Higher AFD = more likely to contribute to fitness differences (and vice versa).
Modified 26Jan21 to streamline

## 1. Load AFD data

```{r load data}
### allele frequency data from script 46.allele.freq.differences.Kgroups
load("./data/pop.af.dat.Rdat")
load("./data/allele.freq.GWAS.snps.Rdata")#afg

### set GWA results
### originally tried to do this with limix results, but limix outputs the beta of the minor allele (with a random sign given to alleles that are at 50%).  This was just too tough to try to "correct" and Ben already set gemma to output the beta of the alt allele, which works much better.  I double checked that this is the case 28Sept20.
### GWAS run in gemma in /groups/nordborg/projects/field_experiments/adaptation_sweden/common.gardens/28.BLUP.GWAS.Rmd
up.gwa.file <- "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds"
```

```{r prep AFD data}
### combine AFD datasets
colnames(pop.af.dat)[1:2] <- c("chrom", "pos")
pos <- do.call(rbind,strsplit(rownames(afg),"_"))
colnames(pos) <- c("chrom", "pos")
pos <- as.data.frame(pos)
afg <- cbind(afg, pos)
af.dat <- merge(pop.af.dat, afg, by=c("chrom", "pos"), all=TRUE)
```

## 2. Add GWAS data and polarize by most frequent local allele

Data ready to go.  Now for each GWAS result, need to load and prep data, then make boxplots.  Write functions for these

```{r data prep fxns}

### function to combine GWAS and AFD data and polarize by "home" and "away" alleles
home.allele <- "ANORTH"
away.allele <- "ASOUTH"
up.gwa.file <- "./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds"

add.gwas <- function(up.gwa.file, home.allele, away.allele){
  load(up.gwa.file) ##variable is named p
  colnames(p)[1:3] <- c("chrom","marker","pos")
  up.dat <- merge(af.dat, p, by=c("chrom", "pos"),all=TRUE)
  
  # get subset where alt allele freq < 0.5 in home population
  home.c <- which(colnames(up.dat)==home.allele)
  away.c <- which(colnames(up.dat)==away.allele)
  maf.dat <- up.dat[up.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat$home.beta <- -(maf.dat$beta)
  
  # do same to subset where alt allele is major allele in home pop
  home.dat <- up.dat[up.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  home.dat$home.beta <- home.dat$beta 
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[-.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  return(out.dat)
}

### extract pheno name from filename
get.p.name <- function(up.gwa.file){
  up.pheno <- unlist(strsplit(up.gwa.file, "/"))
  up.pheno <- up.pheno[grep("rds", up.pheno)]
  up.pheno <- gsub("gemma_lmm_blup_","" ,up.pheno)
  up.pheno <- gsub(".rds","" ,up.pheno)
  return(up.pheno)
}

### make pheno names to short names
short.p.name <- function(p.name){
  p.name <- gsub("ADA","NA",p.name)
  p.name <- gsub("RAM","NM",p.name)
  p.name <- gsub("RAT","SR",p.name)
  p.name <- gsub("ULL","SU",p.name)
  return(p.name)
}
```

## 3.  Effect of most common home allele genome-wide
```{r test home allele beta >0}
gwas.res.files.n <- c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")
gwas.res.files.s <- c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")

home.beta.n <- af.dat[,1:2]
for(up.f in gwas.res.files.n){
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  colnames(up.dat)[colnames(up.dat)=="home.beta"] <- up.pheno
  home.beta.n <- merge(home.beta.n, up.dat[,colnames(up.dat)%in%c("chrom", "pos",up.pheno)], all=TRUE)
}

home.beta.s <- af.dat[,1:2]
for(up.f in gwas.res.files.s){
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  colnames(up.dat)[colnames(up.dat)=="home.beta"] <- up.pheno
  home.beta.s <- merge(home.beta.s, up.dat[,colnames(up.dat)%in%c("chrom", "pos",up.pheno)], all=TRUE)
}

home.beta <- merge(home.beta.n, home.beta.s)

#one-sided t-test mean beta >0
ha.ttest <- apply(home.beta[,3:10],2,function(x){t.test(x, mu = 0, alternative = "greater")})
ha.pval <- lapply(ha.ttest,function(x){x$p.value})
print(ha.pval) #RAM_2011 and all South are significantly >0.  So for all "local adaptation" (plus RAT_2011), home allele on average has positive fitness effect.

##two-sided t-test mean beta
ha.2ttest <- apply(home.beta[,3:10],2,function(x){t.test(x, mu = 0, alternative = "two.sided")})
ha.2pval <- lapply(ha.2ttest,function(x){x$p.value})
print(ha.2pval) #all are sig different from zero, though

## boxplot of beta of home allele values
#reshape data
home.beta.l <- reshape(home.beta,varying=c(3:10), direction="long", timevar="exp",times=colnames(home.beta)[3:10],v.name="beta")
home.beta.l$exp <- gsub("RAM","NM", home.beta.l$exp)
home.beta.l$exp <- gsub("ADA","NA", home.beta.l$exp)
home.beta.l$exp <- gsub("RAT","SR", home.beta.l$exp)
home.beta.l$exp <- gsub("ULL","SU", home.beta.l$exp)
home.beta.l$exp <- factor(home.beta.l$exp,levels=c("NA_2011","NA_2012","NM_2011","NM_2012","SU_2011","SU_2012","SR_2011","SR_2012"))
home.beta.l$region <- substr(home.beta.l$exp,1,1)


home.beta.plot <- ggplot(data = home.beta.l, mapping = aes(x=exp,y=beta,fill=region)) + 
    #geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    #scale_color_manual(values=c("blueviolet")) +
    #geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    geom_violin() +
    scale_fill_manual(values=c("dodgerblue1","springgreen2"), name = "experiment \nregion", labels = c("North", "South")) +
    geom_boxplot(width=0.1, fill="white") +
    labs(x='experiment', y="fitness effect of most common allele in home pop") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=0.5) +
    theme_minimal() +
    stat_summary(geom = 'text', label = c("ns","ns","***","ns","***","***","***","***"), fun = max, vjust = -1)

 plot.name <- "./figures/47.figures/beta.home.allele.violin.plots.jpg"
 ggsave(plot.name, plot=home.beta.plot, width=8, height=6, units="in")

```

## 4. Relationship between home allele frequency and beta
```{r test home allele freq vs beta}

## southern experiments
s.ahome.plots <- as.list(1:4)
s.mw.test <- as.list(1:4)

for(up.fn in 1:4){
  up.f <- gwas.res.files.s[up.fn]
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  up.short.name <- short.p.name(p.name=up.pheno)
  # get AFD bins
  breaks <- seq(0.5,1,0.1)
  # specify interval/bin labels
  tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )  # bucketing values into bins north/south
  up.dat$ahome.bins <- cut(up.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  up.dat$ahome.bins <- factor(up.dat$ahome.bins, levels = tags,ordered = TRUE)
  
   #test difs between allele freq bins
  pt <- pairwise.wilcox.test(up.dat$home.beta, up.dat$ahome.bins, p.adj.method="bh")
  PT <- pt$p.value
  library(rcompanion)
  PT1 = fullPTable(PT)
  comp.lets <- multcompLetters(PT1,compare="<",threshold=0.05,Letters=letters,reversed = FALSE)
  s.mw.test[[up.fn]] <- pt

  haf.boxplot <- ggplot(data = subset(up.dat, !is.na(ahome.bins)), mapping = aes(x=ahome.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='home allele frequency', y="home allele effect") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal() +
    ggtitle(bquote(atop(.(up.short.name), atop(italic(.("home allele = SOUTH")), "")))) +
    stat_summary(geom = 'text', label = comp.lets$Letters, fun = max, vjust = -1)
  #plot.name <- paste("./figures/24.figures/haf.boxplot",up.exp,home.allele,away.allele,"jpg",sep=".")
  #ggsave(plot.name, plot=haf.boxplot, width=4, height=4, units="in")

  s.ahome.plots[[up.fn]] <- haf.boxplot
}

#plot.file <- "./figures/47.figures/home.allele.freq.vs.effect.southern.experiments.jpg"
#  comp.plot <- plot_grid(s.ahome.plots[[1]], s.ahome.plots[[2]], s.ahome.plots[[3]], s.ahome.plots[[4]], ncol = 2, align="v", axis="rl")
#  ggsave(file=plot.file, comp.plot, width=14, height=12, units="in")
  
## northern experiments
n.ahome.plots <- as.list(1:4)
n.mw.test <- as.list(1:4)

for(up.fn in 1:4){
  up.f <- gwas.res.files.n[up.fn]
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  up.short.name <- short.p.name(p.name=up.pheno)
  # get AFD bins
  breaks <- seq(0.5,1,0.1)
  # specify interval/bin labels
  tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )  # bucketing values into bins north/south
  up.dat$ahome.bins <- cut(up.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  up.dat$ahome.bins <- factor(up.dat$ahome.bins, levels = tags,ordered = TRUE)
  
  #test difs between allele freq bins
  pt <- pairwise.wilcox.test(up.dat$home.beta, up.dat$ahome.bins, p.adj.method="bh")
  PT <- pt$p.value
  library(rcompanion)
  PT1 = fullPTable(PT)
  comp.lets <- multcompLetters(PT1,compare="<",threshold=0.05,Letters=letters,reversed = FALSE)
  n.mw.test[[up.fn]] <- pt

  haf.boxplot <- ggplot(data = subset(up.dat, !is.na(ahome.bins)), mapping = aes(x=ahome.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='home allele frequency', y="home allele effect") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal() +
    ggtitle(bquote(atop(.(up.short.name), atop(italic(.("home allele = NORTH")), "")))) +
    stat_summary(geom = 'text', label = comp.lets$Letters, fun = max, vjust = -1)
  #plot.name <- paste("./figures/24.figures/haf.boxplot",up.exp,home.allele,away.allele,"jpg",sep=".")
  #ggsave(plot.name, plot=haf.boxplot, width=4, height=4, units="in")

  n.ahome.plots[[up.fn]] <- haf.boxplot
}


## plot all in one file

plot.file <- "./figures/47.figures/home.allele.freq.vs.effect.all.experiments.jpg"
comp.plot <- plot_grid(n.ahome.plots[[3]], n.ahome.plots[[1]], s.ahome.plots[[3]], s.ahome.plots[[1]], n.ahome.plots[[4]], n.ahome.plots[[2]], s.ahome.plots[[4]], s.ahome.plots[[2]], ncol = 4, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=15, height=12, units="in")
```




## 5. Barplots by AFD bins (polarized by most frequent "local" i.e. N/S allele)

```{r plot betas by allele frequency bins}

#up.dat <- add.gwas(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds", home.allele="AS2", away.allele="AS1")
#up.pheno <- get.p.name(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds")

barplot.home.af.bin <- function(up.dat, up.pheno, home.allele, away.allele, comp.lets){
  nsa.plot <- ggplot(data = up.dat, mapping = aes(x=ha.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='difference between home and away allele frequency', y="home allele effect") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal() +
    stat_summary(geom = 'text', label = comp.lets$Letters, fun = max, vjust = -1)
 nsa.plot <- nsa.plot + ggtitle(up.pheno,subtitle = paste("home allele=",home.allele,", away allele=", away.allele, sep=""))
 #plot.name <- paste("./figures/47.figures/beta.afd.boxplots",up.pheno,home.allele,away.allele,"jpg",sep=".")
 #ggsave(plot.name, plot=nsa.plot, width=8, height=6, units="in")
}

```

```{r plot BLUP GWAS NS results}
## North experiments
n.afd.plots <- as.list(1:4)
n.mw.afd.test <- as.list(1:4)

for(up.fn in 1:4){
  up.f <- gwas.res.files.n[up.fn]
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  up.short.name <- short.p.name(p.name=up.pheno)
  
  #test difs between allele freq bins
  pt <- pairwise.wilcox.test(up.dat$home.beta, up.dat$ha.bins, p.adj.method="bh")
  PT <- pt$p.value
  PT1 = fullPTable(PT)
  comp.lets <- multcompLetters(PT1,compare="<",threshold=0.05,Letters=letters,reversed = FALSE)
  n.mw.afd.test[[up.fn]] <- pt
  
  out.barplot <- barplot.home.af.bin(up.dat=up.dat, up.pheno=up.short.name, home.allele="ANORTH", away.allele="ASOUTH", comp.lets=comp.lets)
  n.afd.plots[[up.fn]] <- out.barplot
}

## South experiments
s.afd.plots <- as.list(1:4)
s.mw.afd.test <- as.list(1:4)

for(up.fn in 1:4){
  up.f <- gwas.res.files.s[up.fn]
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  up.short.name <- short.p.name(p.name=up.pheno)
  
  #test difs between allele freq bins
  pt <- pairwise.wilcox.test(up.dat$home.beta, up.dat$ha.bins, p.adj.method="bh")
  PT <- pt$p.value
  PT1 = fullPTable(PT)
  comp.lets <- multcompLetters(PT1,compare="<",threshold=0.05,Letters=letters,reversed = FALSE)
  s.mw.afd.test[[up.fn]] <- pt
  
  out.barplot <- barplot.home.af.bin(up.dat=up.dat, up.pheno=up.short.name, home.allele="ASOUTH", away.allele="ANORTH", comp.lets=comp.lets)
  s.afd.plots[[up.fn]] <- out.barplot
}


## plot all in one file

plot.file <- "./figures/47.figures/AFD.vs.effect.all.experiments.jpg"
comp.plot <- plot_grid(n.afd.plots[[3]], n.afd.plots[[1]], s.afd.plots[[3]], s.afd.plots[[1]], n.afd.plots[[4]], n.afd.plots[[2]], s.afd.plots[[4]], s.afd.plots[[2]], ncol = 4, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=20, height=12, units="in")
```



## 6.  Contour plots of betas of home allele.  Comparing N vs S experiments
```{r contour plot prep and functions}
## make a big file with all the betas in it
gwas.files <- c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds","./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")

afb.dat <- af.dat
for(up.f in gwas.files){
  #print(up.f)
  load(up.f) ##variable is named p
  up.pheno <- get.p.name(up.gwa.file=up.f)
  colnames(p)[1:3] <- c("chrom","marker","pos")
  p <- p[c(1,3,8)]
  colnames(p)[3] <- up.pheno
  #print(head(p))
  afb.dat <- merge(afb.dat, p, by=c("chrom", "pos"),all=TRUE)
}
#home.allele="ANORTH"
#away.allele="ASOUTH"

polarize.afb <- function(home.allele, away.allele){
## essentially, repolarizing alleles so away allele effect=0 and data file reports home allele effect
  # get subset where alt allele freq < 0.5 in home population
  home.c <- which(colnames(afb.dat)==home.allele)
  away.c <- which(colnames(afb.dat)==away.allele)
  maf.dat <- afb.dat[afb.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat[17:24] <- -maf.dat[17:24]
  
  # do same to subset where alt allele is major allele in home pop (don't need to flip sign)
  home.dat <- afb.dat[afb.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  
  #do home allele frequency bins
  breaks <- seq(0.5,1,0.1)
  tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$ahome.bins <- cut(out.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ahome.bins <- factor(out.dat$ahome.bins, levels = tags,ordered = TRUE)
  
  #do away allele frequency bins
  breaks <- seq(0,1,0.1)
  tags <- c("[.0-.1)","[.1-.2)","[.2-.3)","[.3-.4)","[.4-5)","[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$aaway.bins <- cut(out.dat$aaway, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$aaway.bins <- factor(out.dat$aaway.bins, levels = tags,ordered = TRUE)

  return(out.dat)
}


#h.exp <- "RAM_2011"
#a.exp <- "ULL_2011"
#afb <- afb.s
pw.contour <- function(h.exp,a.exp, afb){
  #calculate centroid
  cx <- mean(afb[,colnames(afb)==h.exp], na.rm=TRUE)
  cy <- mean(afb[,colnames(afb)==a.exp], na.rm=TRUE)
  ##plot
  pw.plot <- ggplot(data=afb, aes(x=get(h.exp), y=get(a.exp))) +
    geom_density_2d() +
    geom_vline(xintercept = 0, color="orange") +
    geom_hline(yintercept = 0, color="orange") +
    xlab(h.exp) +
    ylab(a.exp)+
    theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
    stat_quadrant_counts(aes(label = stat(count))) +
    geom_point(aes(x=cx, y=cy), colour="green", size=2)
}
```

```{R generate single contour plots}

afb.s <- polarize.afb(home.allele="ASOUTH", away.allele="ANORTH")
afb.n <- polarize.afb(home.allele="ANORTH", away.allele="ASOUTH")
## will need to change these as appropriate
## change colnames to easy experiment names
colnames(afb.s) <- gsub("RAM","NM", colnames(afb.n))
colnames(afb.s) <- gsub("ADA","NA", colnames(afb.n))
colnames(afb.s) <- gsub("RAT","SR", colnames(afb.n))
colnames(afb.s) <- gsub("ULL","SU", colnames(afb.n))
colnames(afb.n) <- gsub("RAM","NM", colnames(afb.n))
colnames(afb.n) <- gsub("ADA","NA", colnames(afb.n))
colnames(afb.n) <- gsub("RAT","SR", colnames(afb.n))
colnames(afb.n) <- gsub("ULL","SU", colnames(afb.n))

### need a variable that gives 2 experiments to compare
n.exps <- c("NA_2011", "NA_2012", "NM_2011", "NM_2012")
s.exps <- c("SU_2011", "SU_2012", "SR_2011", "SR_2012")

### North vs South plots
e.combos <- expand.grid(n.exps,s.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
ns.contour.plots <- as.list(1:nrow(e.combos))
for(up.cn in 1:nrow(e.combos)){
  up.c <- e.combos[up.cn,]
  #print(up.c)
  out.plot <- pw.contour(h.exp=as.character(up.c[[1]]),a.exp=as.character(up.c[[2]]), afb=afb.n)
  ns.contour.plots[[up.cn]] <- out.plot
}
plot.file <- "./figures/47.figures/genome.contour.NS.jpg"
comp.plot <- plot_grid(ns.contour.plots[[1]], ns.contour.plots[[2]], ns.contour.plots[[3]], ns.contour.plots[[4]], ns.contour.plots[[5]], ns.contour.plots[[6]], ns.contour.plots[[7]], ns.contour.plots[[8]],ns.contour.plots[[9]], ns.contour.plots[[10]], ns.contour.plots[[11]], ns.contour.plots[[12]], ns.contour.plots[[13]], ns.contour.plots[[14]], ns.contour.plots[[15]], ns.contour.plots[[16]], ncol = 4, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=20, height=20, units="in")

### South vs North plots
e.combos <- expand.grid(s.exps,n.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
sn.contour.plots <- as.list(1:nrow(e.combos))
for(up.cn in 1:nrow(e.combos)){
  up.c <- e.combos[up.cn,]
  #print(up.c)
  out.plot <- pw.contour(h.exp=as.character(up.c[[1]]),a.exp=as.character(up.c[[2]]), afb=afb.s)
  sn.contour.plots[[up.cn]] <- out.plot
}
plot.file <- "./figures/47.figures/genome.contour.SN.jpg"
comp.plot <- plot_grid(sn.contour.plots[[1]], sn.contour.plots[[2]], sn.contour.plots[[3]], sn.contour.plots[[4]], sn.contour.plots[[5]], sn.contour.plots[[6]], sn.contour.plots[[7]], sn.contour.plots[[8]],sn.contour.plots[[9]], sn.contour.plots[[10]], sn.contour.plots[[11]], sn.contour.plots[[12]], sn.contour.plots[[13]], sn.contour.plots[[14]], sn.contour.plots[[15]], sn.contour.plots[[16]], ncol = 4, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=20, height=20, units="in")


### North vs North plots
e.combos <- expand.grid(n.exps,n.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
nn.contour.plots <- as.list(1:nrow(e.combos))
for(up.cn in 1:nrow(e.combos)){
  up.c <- e.combos[up.cn,]
  #print(up.c)
  out.plot <- pw.contour(h.exp=as.character(up.c[[1]]),a.exp=as.character(up.c[[2]]), afb=afb.n)
  nn.contour.plots[[up.cn]] <- out.plot
}
plot.file <- "./figures/47.figures/genome.contour.NN.jpg"
comp.plot <- plot_grid(nn.contour.plots[[1]], nn.contour.plots[[2]], nn.contour.plots[[3]], nn.contour.plots[[4]], nn.contour.plots[[5]], nn.contour.plots[[6]], nn.contour.plots[[7]], nn.contour.plots[[8]],nn.contour.plots[[9]], nn.contour.plots[[10]], nn.contour.plots[[11]], nn.contour.plots[[12]], nn.contour.plots[[13]], nn.contour.plots[[14]], nn.contour.plots[[15]], nn.contour.plots[[16]], ncol = 4, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=20, height=20, units="in")


### South vs South plots
e.combos <- expand.grid(s.exps,s.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
ss.contour.plots <- as.list(1:nrow(e.combos))
for(up.cn in 1:nrow(e.combos)){
  up.c <- e.combos[up.cn,]
  #print(up.c)
  out.plot <- pw.contour(h.exp=as.character(up.c[[1]]),a.exp=as.character(up.c[[2]]), afb=afb.s)
  ss.contour.plots[[up.cn]] <- out.plot
}
plot.file <- "./figures/47.figures/genome.contour.SS.jpg"
comp.plot <- plot_grid(ss.contour.plots[[1]], ss.contour.plots[[2]], ss.contour.plots[[3]], ss.contour.plots[[4]], ss.contour.plots[[5]], ss.contour.plots[[6]], ss.contour.plots[[7]], ss.contour.plots[[8]],ss.contour.plots[[9]], ss.contour.plots[[10]], ss.contour.plots[[11]], ss.contour.plots[[12]], ss.contour.plots[[13]], ss.contour.plots[[14]], ss.contour.plots[[15]], ss.contour.plots[[16]], ncol = 4, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=20, height=20, units="in")

```

## 7.  Centroid positions in different AFD bins - pairwise comparisons

produce a sample plot of contours by AFD with centroids marked
Idea is to show how plots that follow are generated

```{r contour AFD with centroids example}

h.exp <- "SU_2011"
a.exp <- "NM_2011"
up.dat <- afb.s

#generate centroids per bin
maxb <- max(abs(up.dat[,colnames(up.dat)%in%c(h.exp, a.exp)]),na.rm=TRUE)
hmean <- aggregate(get(h.exp)~ha.bins, data=up.dat, mean, na.action = na.omit)
amean <- aggregate(get(a.exp)~ha.bins, data=up.dat, mean, na.action = na.omit)
bmeans <- merge(hmean, amean)
colnames(bmeans)[2:3] <- c("h.mean", "a.mean")

afd.cen.plot <- ggplot(data=up.dat, aes(x=get(h.exp), y=get(a.exp))) +
    geom_density_2d() +
    facet_wrap(~ha.bins,nrow=2) +
    geom_vline(xintercept = 0, color="orange") +
    geom_hline(yintercept = 0, color="orange") +
    xlab(h.exp) +
    ylab(a.exp)+
    theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
    stat_quadrant_counts(aes(label = stat(count))) +
    geom_point(data=bmeans, aes(x=h.mean, y=a.mean, color=ha.bins), size=4) +
    theme_bw() +
    scale_color_viridis(discrete = TRUE, option = "D", direction = 1, begin=0.1, end=0.99) +
    labs(color = "allele frequency\ndifference bin")

pdf(file="./figures/47.figures/centroids.byAFD.ULL11.RAM11.pdf", width=10, height=5)
print(afd.cen.plot)
dev.off()

```


combine centroids by AFD into one plot (without the contours of the distribution)
    
```{r centroids by home-away AFD functions}    
#h.exp <- "SU_2011"
#a.exp <- "NM_2011"
#up.dat <- afb.s

cen.by.afd.plot <- function(h.exp, a.exp, up.dat){
  #generate centroids per bin
  maxb <- max(abs(up.dat[,colnames(up.dat)%in%c(h.exp, a.exp)]),na.rm=TRUE)
  hmean <- aggregate(get(h.exp)~ha.bins, data=up.dat, mean, na.action=na.omit)
  amean <- aggregate(get(a.exp)~ha.bins, data=up.dat, mean, na.action=na.omit)
  bmeans <- merge(hmean, amean)
  colnames(bmeans)[2:3] <- c("h.mean", "a.mean")
  
  out.plot <- ggplot(bmeans, aes(x=h.mean, y=a.mean, color=ha.bins)) +
    geom_point(size=2.5) +
    geom_vline(xintercept = 0, color="orange") +
    geom_hline(yintercept = 0, color="orange") +
    scale_color_viridis(discrete = TRUE, option = "D", direction = 1, begin=0.1, end=0.99) +
    xlab(h.exp) +
    ylab(a.exp) +
    labs(color = "allele frequency\ndifference bin")+ 
    theme_bw()
  
  return(out.plot)
}
```

```{r centroids by AFD NS plots}
e.combos <- expand.grid(n.exps,s.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
ns.afd.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.afd.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.n)
  ns.afd.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.byAFD.NS.jpg"
comp.plot <- plot_grid(
  ns.afd.centroid.plots[[1]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[3]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[2]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[4]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[9]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[11]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[10]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[12]]+ theme(legend.position="none"),
  ns.afd.centroid.plots[[5]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[7]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[6]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[8]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[13]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[15]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[14]]+ theme(legend.position="none"), 
  ns.afd.centroid.plots[[16]]+ theme(legend.position="none"), 
  ncol = 4, align="v", axis="rl")

legend_b <- get_legend(ns.afd.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=12, height=10, units="in")
```


```{r centroids by AFD SN plots}
e.combos <- expand.grid(s.exps,n.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
sn.afd.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.afd.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.s)
  sn.afd.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.byAFD.SN.jpg"
comp.plot <- plot_grid(
  sn.afd.centroid.plots[[1]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[3]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[2]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[4]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[9]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[11]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[10]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[12]]+ theme(legend.position="none"),
  sn.afd.centroid.plots[[5]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[7]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[6]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[8]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[13]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[15]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[14]]+ theme(legend.position="none"), 
  sn.afd.centroid.plots[[16]]+ theme(legend.position="none"), 
  ncol = 4, align="v", axis="rl")

legend_b <- get_legend(ns.afd.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=12, height=10, units="in")
```

```{r centroids by AFD NN plots}
e.combos <- expand.grid(n.exps,n.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
nn.afd.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.afd.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.n)
  nn.afd.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.byAFD.NN.jpg"
comp.plot <- plot_grid(
  nn.afd.centroid.plots[[1]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[3]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[2]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[4]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[9]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[11]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[10]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[12]]+ theme(legend.position="none"),
  nn.afd.centroid.plots[[5]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[7]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[6]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[8]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[13]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[15]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[14]]+ theme(legend.position="none"), 
  nn.afd.centroid.plots[[16]]+ theme(legend.position="none"), 
  ncol = 4, align="v", axis="rl")

legend_b <- get_legend(ns.afd.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=12, height=10, units="in")
```

```{r centroids by AFD SS plots}
e.combos <- expand.grid(s.exps,s.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
ss.afd.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.afd.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.s)
  ss.afd.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.byAFD.SS.jpg"
comp.plot <- plot_grid(
  ss.afd.centroid.plots[[1]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[3]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[2]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[4]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[9]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[11]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[10]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[12]]+ theme(legend.position="none"),
  ss.afd.centroid.plots[[5]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[7]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[6]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[8]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[13]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[15]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[14]]+ theme(legend.position="none"), 
  ss.afd.centroid.plots[[16]]+ theme(legend.position="none"), 
  ncol = 4, align="v", axis="rl")

legend_b <- get_legend(ns.afd.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=12, height=10, units="in")
```




########################################
############# not rewritten yet 15Feb21
########################################

## 8.  Centroid positions in home frequency bins colored by away frequency bins
```{r centroid by both AF bins - demo plot}
up.dat <- afb.s
h.exp <- "SU_2011"
a.exp <- "NM_2011"

hmean <- aggregate(get(h.exp)~ahome.bins+aaway.bins, data=up.dat, mean)
amean <- aggregate(get(a.exp)~ahome.bins+aaway.bins, data=up.dat, mean)
bmeans <- merge(hmean, amean)
colnames(bmeans)[3:4] <- c(h.exp, a.exp)

both.cen.plot <- ggplot(data=up.dat, aes(x=get(h.exp), y=get(a.exp))) +
    geom_density_2d() +
    facet_grid(aaway.bins~ahome.bins) +
    geom_vline(xintercept = 0, color="orange") +
    geom_hline(yintercept = 0, color="orange") +
    xlab(h.exp) +
    ylab(a.exp)+
    theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
    stat_quadrant_counts(aes(label = stat(count))) +
    geom_point(data=bmeans, aes(x=get(h.exp), y=get(a.exp), color=aaway.bins), size=4) +
    theme_bw() +
    scale_color_viridis(discrete = TRUE, option = "D", direction = 1, begin=0.1, end=0.99) +
    labs(color = "away allele\nfrequency")

pdf(file="./figures/47.figures/centroids.by.bothAF.ULL11.RAM11.pdf", width=10, height=10)
print(both.cen.plot)
dev.off()

```




```{r centroid by both AF bins fxns}
up.dat <- afb.s
h.exp <- "SU_2011"
a.exp <- "NM_2012"

cen.by.bothAF.plot <- function(h.exp, a.exp, up.dat){
  hmean <- aggregate(get(h.exp)~ahome.bins+aaway.bins, data=up.dat, mean)
  amean <- aggregate(get(a.exp)~ahome.bins+aaway.bins, data=up.dat, mean)
  bmeans <- merge(hmean, amean)
  colnames(bmeans)[3:4] <- c(h.exp, a.exp)

  bothAF.cen.plot <- ggplot(data=bmeans, aes(x=get(h.exp), y=get(a.exp), color=aaway.bins)) +
      geom_point(size=4) +
      facet_wrap(~ahome.bins, nrow=1) +
      geom_vline(xintercept = 0, color="orange") +
      geom_hline(yintercept = 0, color="orange") +
      xlab(h.exp) +
      ylab(a.exp) +
      #theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
      #stat_quadrant_counts(aes(label = stat(count))) +
      theme_bw() +
      scale_color_viridis(discrete = TRUE, option = "D", direction = 1, begin=0.1, end=0.99) +
      #labs(color = "allele frequency\nin away\npopulation")
      labs(color = "allele frequency in away population")
  
    return(bothAF.cen.plot)
}
```

```{r centroid by both AF bins NS plots}
e.combos <- expand.grid(n.exps,s.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
ns.both.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.bothAF.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.n)
  ns.both.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.bothAF.NS.jpg"
comp.plot <- plot_grid(
  ns.both.centroid.plots[[1]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[3]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[2]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[4]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[9]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[11]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[10]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[12]]+ theme(legend.position="none"),
  ns.both.centroid.plots[[5]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[7]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[6]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[8]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[13]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[15]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[14]]+ theme(legend.position="none"), 
  ns.both.centroid.plots[[16]]+ theme(legend.position="none"), 
  ncol = 4, align="v", axis="rl")

legend_b <- get_legend(ns.both.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=20, height=10, units="in")

```

```{r centroid by both AF bins SN plots}
e.combos <- expand.grid(s.exps,n.exps)
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
sn.both.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.bothAF.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.s)
  sn.both.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.bothAF.SN.jpg"
comp.plot <- plot_grid(
  sn.both.centroid.plots[[1]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[3]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[2]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[4]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[9]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[11]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[10]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[12]]+ theme(legend.position="none"),
  sn.both.centroid.plots[[5]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[7]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[6]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[8]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[13]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[15]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[14]]+ theme(legend.position="none"), 
  sn.both.centroid.plots[[16]]+ theme(legend.position="none"), 
  ncol = 4, align="v", axis="rl")

legend_b <- get_legend(sn.both.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=20, height=10, units="in")

```

```{r centroid by both AF bins NN plots}
e.combos <- t(combn(n.exps,2))
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)

nn.both.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.bothAF.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.n)
  nn.both.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.bothAF.NN.jpg"
comp.plot <- plot_grid(
  nn.both.centroid.plots[[1]]+ theme(legend.position="none"), 
  nn.both.centroid.plots[[2]]+ theme(legend.position="none"), 
  nn.both.centroid.plots[[3]]+ theme(legend.position="none"), 
  nn.both.centroid.plots[[4]]+ theme(legend.position="none"), 
  nn.both.centroid.plots[[5]]+ theme(legend.position="none"), 
  nn.both.centroid.plots[[6]]+ theme(legend.position="none"), 
  ncol = 3, align="v", axis="rl")

legend_b <- get_legend(nn.both.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=15, height=5, units="in")

```

```{r centroid by both AF bins SS plots}
e.combos <- t(combn(s.exps,2))
colnames(e.combos) <- c("focal.exp","comp.exp")
e.combos <- as.data.frame(e.combos)
ss.both.centroid.plots <- as.list(1:nrow(e.combos))

#generate plots
for(up in 1:nrow(e.combos)){
  up.plot <- cen.by.bothAF.plot(h.exp=as.character(e.combos[up,1]), a.exp=as.character(e.combos[up,2]), up.dat=afb.n)
  ss.both.centroid.plots[[up]] <- up.plot
}

#save plots
plot.file <- "./figures/47.figures/centroids.bothAF.SS.jpg"
comp.plot <- plot_grid(
  ss.both.centroid.plots[[1]]+ theme(legend.position="none"), 
  ss.both.centroid.plots[[2]]+ theme(legend.position="none"), 
  ss.both.centroid.plots[[3]]+ theme(legend.position="none"), 
  ss.both.centroid.plots[[4]]+ theme(legend.position="none"), 
  ss.both.centroid.plots[[5]]+ theme(legend.position="none"), 
  ss.both.centroid.plots[[6]]+ theme(legend.position="none"), 
  ncol = 3, align="v", axis="rl")

legend_b <- get_legend(ss.both.centroid.plots[[1]] + 
    guides(color = guide_legend(nrow = 1)) +
    theme(legend.position = "bottom"))

out.plot <- plot_grid(comp.plot, legend_b, ncol = 1, rel_heights = c(1, .1))

ggsave(file=plot.file, out.plot, width=15, height=5, units="in")

```


### 9. pairwise betas, by homeAF and awayAF, box sizes by quadrant
```{r scatterplots beta by all AF combos, eval=FALSE}

# this does scatterplots by all combos
pw.both.bins.plot <- ggplot(data=out.dat, aes(x=get(up.exp), y=get(comp.exp))) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  facet_grid(aaway.bins~ahome.bins) +
  geom_vline(xintercept = 0, color="orange") +
  geom_hline(yintercept = 0, color="orange") +
  xlab(up.exp) +
  ylab(comp.exp)+
  theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
  stat_quadrant_counts(aes(label = stat(count))) +
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), ""))))

```


```{r percentage by quadrats for all AF combos fxn}

af.combo.plot <- function(home.allele, away.allele, up.exp, comp.exp) {

  ## essentially, repolarizing alleles so away allele effect=0 and data file reports home allele effect
  # get subset where alt allele freq < 0.5 in home population
  home.c <- which(colnames(afb.dat)==home.allele)
  away.c <- which(colnames(afb.dat)==away.allele)
  maf.dat <- afb.dat[afb.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat[17:24] <- -maf.dat[17:24]
  
  # do same to subset where alt allele is major allele in home pop (don't need to flip sign)
  home.dat <- afb.dat[afb.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  
  #do home allele frequency bins
  breaks <- seq(0.5,1,0.1)
  tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$ahome.bins <- cut(out.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ahome.bins <- factor(out.dat$ahome.bins, levels = tags,ordered = TRUE)
  
  #do away allele frequency bins
  breaks <- seq(0,1,0.1)
  tags <- c("[.0-.1)","[.1-.2)","[.2-.3)","[.3-.4)","[.4-5)","[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$aaway.bins <- cut(out.dat$aaway, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$aaway.bins <- factor(out.dat$aaway.bins, levels = tags,ordered = TRUE)
  
  ## need to reshape the data and calculate quadrant numbers
  ue.col <- which(colnames(out.dat)==up.exp)
  ce.col <- which(colnames(out.dat)==comp.exp)
  combos <- expand.grid(levels(out.dat$ahome.bins), levels(out.dat$aaway.bins))
  c.table <- apply(combos,1, function(x){
    ugh <- out.dat[out.dat$ahome.bins==x[1] & out.dat$aaway.bins==x[2],]
    ugh.t <- table(ugh[,ue.col]>0, ugh[,ce.col]>0)
    ugh.u <- as.numeric(ugh.t)
    return(ugh.u)
  })
  c.table <- t(c.table)
  c.table <- as.data.frame(c.table)
  colnames(c.table) <- c("NN","PN","NP","PP") ## N=negative, P=positive, first is home, second is away betas
  c.table <- cbind(c.table,combos)
  colnames(c.table)[5:6] <- c("ahome.bins", "aaway.bins")

  ## turn this into proportions
  c.table$totals <- apply(c.table[,1:4],1, sum)
  c.table.l <- gather(c.table, quadrant, snp.no, NN:PP, factor_key=TRUE)
  c.table.l$snp.p <- c.table.l$snp.no/c.table.l$totals

  cpd <- c.table.l %>% 
    mutate(xmin = if_else(quadrant %in% c("NN","NP"), -sqrt(snp.p), 0),
         xmax = if_else(quadrant %in% c("PP", "PN"), sqrt(snp.p), 0),
         ymin = if_else(quadrant %in% c("NN","PN"), -sqrt(snp.p), 0),
         ymax = if_else(quadrant %in% c("PP","NP"), sqrt(snp.p), 0))

  combo.plot <- ggplot(cpd) + 
    geom_rect(aes(xmin=xmin, xmax=xmax, ymin=ymin, ymax=ymax), fill = "grey") +
    geom_text(aes(x = xmin + 0.5*sqrt(snp.p), y = ymin +0.5*sqrt(snp.p), label = formatC(round(snp.p*100,2), digits=0,format="f")),color = "blue", size=3) +
    facet_grid(aaway.bins~ahome.bins,switch="both") +
    coord_equal() +
    geom_hline(yintercept = 0) +
    geom_vline(xintercept = 0) +
    labs(title = "Percent SNPs per beta quadrant", x=paste("home AF/ ", up.exp," betas", sep=""), y=paste("away AF/ ", comp.exp, "  betas", sep="")) +
    theme_minimal() +
    theme(axis.text = element_blank(),
        panel.grid = element_blank(),
        plot.title = element_text(color="grey40", face="bold", size=10, hjust = 0.5))
  
  plot.file <- paste("./figures/47.figures/quadrant.af.",up.exp,".",comp.exp,".pdf", sep="")
  ggsave(file=plot.file, combo.plot, width=8, height=12, units="in")
}

#af.combo.plot(out.dat=out.dat, up.exp="RAM_2011", comp.exp="ULL_2011")
```



```{r percentage by quadrats for all AF combos generate, eval=FALSE}
### ULL_2011
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="ULL_2011", comp.exp="RAM_2011")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="ULL_2011", comp.exp="ADA_2011")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="ULL_2011", comp.exp="RAT_2011")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="ULL_2011", comp.exp="ULL_2012")

### RAT_2011
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="RAT_2011", comp.exp="RAM_2011")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="RAT_2011", comp.exp="ADA_2011")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="RAT_2011", comp.exp="RAT_2012")

### ULL_2012
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="ULL_2012", comp.exp="RAM_2012")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="ULL_2012", comp.exp="ADA_2012")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="ULL_2012", comp.exp="RAT_2012")

### RAT_2012
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="RAT_2012", comp.exp="RAM_2012")
af.combo.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp="RAT_2012", comp.exp="ADA_2012")

### RAM_2011
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="RAM_2011", comp.exp="ULL_2011")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="RAM_2011", comp.exp="RAT_2011")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="RAM_2011", comp.exp="ADA_2011")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="RAM_2011", comp.exp="RAM_2012")

### ADA_2011
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="ADA_2011", comp.exp="ULL_2011")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="ADA_2011", comp.exp="RAT_2011")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="ADA_2011", comp.exp="ADA_2012")

### RAM_2012
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="RAM_2012", comp.exp="ULL_2012")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="RAM_2012", comp.exp="RAT_2012")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="RAM_2012", comp.exp="ADA_2012")

### ADA_2012
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="ADA_2012", comp.exp="ULL_2012")
af.combo.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp="ADA_2012", comp.exp="RAT_2012")

```







###########################################
### Old stuff beyond here
##########################################

## 4. Barplots for Southern experiments by AFD bins (polarized by most frequent allele in S1 or S2 versus N or other S allele)

```{r plot BLUP GWAS S1/S2 results}

#S1vsS2, S1 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AS1", away.allele="AS2")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AS1", away.allele="AS2")
}

#S1vsS2, S2 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AS2", away.allele="AS1")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AS2", away.allele="AS1")
}

#S1vsN, S1 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AS1", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AS1", away.allele="ANORTH")
}

#S2vsN, S2 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AS2", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AS2", away.allele="ANORTH")
}
```

## 5. Barplots for Northern experiments by AFD bins (polarized by most frequent allele in N1 or N2 versus S or other N allele)

```{r plot BLUP GWAS N1/N2 results}

#N1vsN2, N1 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AN1", away.allele="AN2")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AN1", away.allele="AN2")
}

##N1vsN2, N2 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AN2", away.allele="AN1")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AN2", away.allele="AN1")
}

#N1vsS, N1 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AN1", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AN1", away.allele="ASOUTH")
}

#N2vsS, N2 as home
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")){
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="AN2", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.home.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="AN2", away.allele="ASOUTH")
}
```

## 6. Barplots by AFD bins (polarized by most frequent "away" i.e. N/S allele)

```{r data prep away allele fxns}

### function to combine GWAS and AFD data and polarize by "home" and "away" alleles
home.allele <- "ANORTH"
away.allele <- "ASOUTH"
up.gwa.file <- "./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds"

add.gwas.af <- function(up.gwa.file, home.allele, away.allele){
  load(up.gwa.file) ##variable is named p
  colnames(p)[1:3] <- c("chrom","marker","pos")
  up.dat <- merge(af.dat, p, by=c("chrom", "pos"),all=TRUE)
  
  # get subset where alt allele freq < 0.5 in away population - will have to switch these
  home.c <- which(colnames(up.dat)==home.allele)
  away.c <- which(colnames(up.dat)==away.allele)
  maf.dat <- up.dat[up.dat[,away.c]< 0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat$home.beta <- -(maf.dat$beta)
  
  # do same to subset where alt allele is major allele in away pop
  home.dat <- up.dat[up.dat[,away.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  home.dat$home.beta <- home.dat$beta 
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-1,0.5,0.15)
  # specify interval/bin labels
  tags <- c("-[1-.85)","-[.85-.7)", "-[.7-.55)", "-[.55-.4)", "-[-.4-.25)","-[.25-.1)", "[-.1-.05)","[.05-.2)", "[.2-35)","[.35-.5)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  return(out.dat)
}
```

```{r plot betas by allele frequency bins by most freq away allele}

up.dat <- add.gwas.af(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds", home.allele="ASOUTH", away.allele="ANORTH")
up.pheno <- get.p.name(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds")

barplot.away.af.bin <- function(up.dat, up.pheno, home.allele, away.allele){
  nsa.plot <- ggplot(data = up.dat, mapping = aes(x=ha.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='AFD home-away (polarized by most common away allele)', y="beta of most common allele in away pop") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal()
 nsa.plot <- nsa.plot + ggtitle(up.pheno,subtitle = paste("home allele=",home.allele,", away allele=", away.allele, sep=""))
 plot.name <- paste("./figures/47.figures/beta.afd.away.boxplots",up.pheno,home.allele,away.allele,"jpg",sep=".")
 ggsave(plot.name, plot=nsa.plot, width=8, height=6, units="in")
}

#barplot.away.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="ASOUTH", away.allele="ANORTH")
```

```{r plot BLUP GWAS NS results by minor allele, eval=FALSE}
## North sites
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")){
  print(up.f)
  up.dat <- add.gwas.af(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.away.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="ANORTH", away.allele="ASOUTH")
}

## South sites
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")){
  print(up.f)
  up.dat <- add.gwas.af(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  barplot.away.af.bin(up.dat=up.dat, up.pheno=up.pheno, home.allele="ASOUTH", away.allele="ANORTH")
}
```

## 7. Simple scatterplots of beta by home allele frequency 

```{r fxn scatterplot beta by AF}
#up.dat <- add.gwas(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", home.allele="ASOUTH", away.allele="ANORTH")
#up.pheno <- get.p.name(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds")

plot.af.scatter <- function(up.dat, up.pheno){
  pdf(file=paste("./figures/47.figures/", up.pheno,".AFD.scatterplots.pdf",sep=""), width=10, height=4)
  par(mfcol=c(1,3))
  smoothScatter(up.dat$home.beta~up.dat$ahome, xlab="home allele freq", ylab="home beta", main=up.pheno)
  abline(h=0,col="red", lty=2)
  abline(lm(home.beta~ahome, data=up.dat), col="green")

  smoothScatter(up.dat$home.beta~up.dat$ha.afd, xlab="AFD polarized by most frequent home allele", ylab="home beta", main=up.pheno)
  abline(h=0,col="red", lty=2)
  abline(lm(home.beta~ha.afd, data=up.dat), col="green")

  smoothScatter(up.dat$ha.afd, up.dat$ahome, xlab="home AFD", ylab="home AF", main=up.pheno)
  dev.off()
}

#plot.af.scatter(up.dat=up.dat, up.pheno=up.pheno)
```

```{R draw af scatterplots, eval=FALSE}
## North sites
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  plot.af.scatter(up.dat=up.dat, up.pheno=up.pheno)
}

## South sites
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")){
  print(up.f)
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  plot.af.scatter(up.dat=up.dat, up.pheno=up.pheno)
}
```

## 8. Simple scatterplots of beta by away allele frequency 

```{r fxn scatterplot beta by AF}
#up.dat <- add.gwas.af(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", home.allele="ASOUTH", away.allele="ANORTH")
#up.pheno <- get.p.name(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds")

plot.af.away.scatter <- function(up.dat, up.pheno){
  pdf(file=paste("./figures/47.figures/", up.pheno,".AFD.away.scatterplots.pdf",sep=""), width=10, height=4)
  par(mfcol=c(1,3))
  smoothScatter(up.dat$home.beta~up.dat$aaway, xlab="away allele freq", ylab="home beta", main=up.pheno)
  abline(h=0,col="red", lty=2)
  abline(lm(home.beta~aaway, data=up.dat), col="green")

  smoothScatter(up.dat$home.beta~up.dat$ha.afd, xlab="AFD home-away polarized by most frequent away allele", ylab="home beta", main=up.pheno)
  abline(h=0,col="red", lty=2)
  abline(lm(home.beta~ha.afd, data=up.dat), col="green")

  smoothScatter(up.dat$ha.afd, up.dat$ahome, xlab="home-away AFD", ylab="away AF", main=up.pheno)
  dev.off()
}

#plot.af.away.scatter(up.dat=up.dat, up.pheno=up.pheno)
```

```{R draw AF away scatterplots, eval=FALSE}
## North sites
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")){
  print(up.f)
  up.dat <- add.gwas.af(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  plot.af.away.scatter(up.dat=up.dat, up.pheno=up.pheno)
}

## South sites
for(up.f in c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")){
  print(up.f)
  up.dat <- add.gwas.af(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  plot.af.away.scatter(up.dat=up.dat, up.pheno=up.pheno)
}
```
## 9. Pairwise beta scatterplots polarized by home allele
```{r pairwise beta data prep}
## make a big file with all the betas in it
gwas.files <- c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds","./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")

afb.dat <- af.dat
for(up.f in gwas.files){
  #print(up.f)
  load(up.f) ##variable is named p
  up.pheno <- get.p.name(up.gwa.file=up.f)
  colnames(p)[1:3] <- c("chrom","marker","pos")
  p <- p[c(1,3,8)]
  colnames(p)[3] <- up.pheno
  #print(head(p))
  afb.dat <- merge(afb.dat, p, by=c("chrom", "pos"),all=TRUE)
}
```

```{r pairwise beta functions}
up.exp <- "RAM_2011"
comp.exp <- "ULL_2011"
home.allele <- "ANORTH"
away.allele <- "ASOUTH"


beta.by.bin.plots <- function(up.exp, comp.exp, home.allele, away.allele){
  ## essentially, repolarizing alleles so away allele effect=0 and data file reports home allele effect
  # get subset where alt allele freq < 0.5 in home population
  home.c <- which(colnames(afb.dat)==home.allele)
  away.c <- which(colnames(afb.dat)==away.allele)
  maf.dat <- afb.dat[afb.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat[17:24] <- -maf.dat[17:24]
  
  # do same to subset where alt allele is major allele in home pop (don't need to flip sign)
  home.dat <- afb.dat[afb.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  
  #do home allele frequency bins
  breaks <- seq(0.5,1,0.1)
  tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$ahome.bins <- cut(out.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ahome.bins <- factor(out.dat$ahome.bins, levels = tags,ordered = TRUE)
  
  #do away allele frequency bins
  breaks <- seq(0,1,0.1)
  tags <- c("[.0-.1)","[.1-.2)","[.2-.3)","[.3-.4)","[.4-5)","[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$aaway.bins <- cut(out.dat$aaway, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$aaway.bins <- factor(out.dat$aaway.bins, levels = tags,ordered = TRUE)
  
  plot.title <- paste(up.exp, comp.exp, sep="-")
  plot.subtitle <- paste("home=",home.allele, " away=",away.allele, sep="")

pw.bins.plot <- ggplot(data=out.dat, aes(x=get(up.exp), y=get(comp.exp))) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  facet_grid(~ha.bins) +
  geom_vline(xintercept = 0, color="orange") +
  geom_hline(yintercept = 0, color="orange") +
  xlab(up.exp) +
  ylab(comp.exp)+
  theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
  stat_quadrant_counts(aes(label = stat(count))) +
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), "")))) 


pw.plot <- ggplot(data=out.dat, aes(x=get(up.exp), y=get(comp.exp))) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  geom_vline(xintercept = 0, color="orange") +
  geom_hline(yintercept = 0, color="orange") +
  xlab(up.exp) +
  ylab(comp.exp)+
  theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
  stat_quadrant_counts(aes(label = stat(count))) +
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), "")))) 

c.fit <- matrix(nrow=length(levels(out.dat$ha.bins)), ncol=4)
for(up in 1:length(levels(out.dat$ha.bins))){
  up.b <- levels(out.dat$ha.bins)[up]
  up.dat <- out.dat[out.dat$ha.bins==up.b,]
  ue.col <- which(colnames(up.dat)==up.exp)
  ce.col <- which(colnames(up.dat)==comp.exp)
  sh1 <- sum(up.dat$ahome*up.dat[,ue.col], na.rm=TRUE)
  sa1 <- sum(up.dat$aaway*up.dat[,ue.col], na.rm=TRUE)
  sh2 <- sum(up.dat$ahome*up.dat[,ce.col], na.rm=TRUE)
  sa2 <- sum(up.dat$aaway*up.dat[,ce.col], na.rm=TRUE)
  out.fit <- c(sh1, sa1 ,sh2,sa2)
  c.fit[up,] <-out.fit
}
colnames(c.fit) <- c("sh1", "sa1","sh2","sa2")
rownames(c.fit) <- levels(out.dat$ha.bins)
c.fit <- as.data.frame(c.fit)
c.fit$bin <- rownames(c.fit)
c.fit.l <- gather(c.fit, pop.exp, c.fitness, sh1, sa1, sh2, sa2)
c.fit.l$pop.exp <- factor(c.fit.l$pop.exp, levels=c("sh1","sa1","sh2","sa2"))
c.fit.l$bin <- factor(c.fit.l$bin, levels=c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)"))

#number of SNPs in each bin
hen <- out.dat[is.na(out.dat[,ue.col])==FALSE,]
ht <- table(hen$ha.bins)
ht <- as.matrix(ht)
aen <- out.dat[is.na(out.dat[,ce.col])==FALSE,]
at <- table(aen$ha.bins)
at <- as.matrix(at)
sn.index <- cbind(ht,ht,at,at)
colnames(sn.index) <- c("sh1","sa1","sh2","sa2")
sn.index <- as.data.frame(sn.index)
sn.index$bin <- rownames(sn.index)
sn.index.l <- gather(sn.index, pop.exp, snp.no, sh1, sa1, sh2, sa2)

c.fit.l <- merge(c.fit.l, sn.index.l, by=c("pop.exp", "bin"))
c.fit.l$fit.per.snp <- c.fit.l$c.fitness/c.fit.l$snp.no


f.plot <-ggplot(c.fit.l, aes(x=bin, y=c.fitness, group=pop.exp)) + 
  geom_line(aes(color=pop.exp)) + 
  geom_point(aes(color=pop.exp, size=abs(fit.per.snp))) + 
  xlab("AFD bin") + 
  ylab("population fitness of bin") + 
  #scale_color_discrete(name = "pop/EXP", labels = c("home/HOME", "away/HOME", "home/AWAY","away/AWAY")) +
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), "")))) +
  geom_hline(yintercept = 0, color="grey50", linetype="dashed") +
  scale_color_manual(values=c("red", "pink2", "dodgerblue","lightblue2"),name = "EXP/pop", labels=c("HOME/home", "HOME/away", "AWAY/home","AWAY/away")) +
  scale_size(name="mean effect per SNP\n(absolute value)") +
  theme_linedraw()  

#generate plots
plot.file <- paste("./figures/47.figures/",up.exp,"_",comp.exp,"_",home.allele,"_",away.allele,"_beta_pairwise_by_AFD_bins.pdf",sep="")
#pdf(file=plot.file, width=14, height=8)
comp.plot <- plot_grid(pw.bins.plot, f.plot, ncol = 1, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=14, height=8, units="in")
#dev.off()
}


up.exp <- "RAM_2011"
comp.exp <- "ULL_2011"
home.allele <- "ANORTH"
away.allele <- "ASOUTH"


#beta.by.bin.plots(up.exp="ULL_2011", comp.exp="RAM_2011", home.allele="ASOUTH", away.allele="ANORTH")
```


```{r generate pairwise beta plots, eval=FALSE}
#ULL_2011
beta.by.bin.plots(up.exp="ULL_2011", comp.exp="RAM_2011", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="ULL_2011", comp.exp="ADA_2011", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="ULL_2011", comp.exp="RAT_2011", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="ULL_2011", comp.exp="ULL_2012", home.allele="ASOUTH", away.allele="ANORTH")

#ULL_2012
beta.by.bin.plots(up.exp="ULL_2012", comp.exp="RAM_2012", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="ULL_2012", comp.exp="ADA_2012", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="ULL_2012", comp.exp="RAT_2012", home.allele="ASOUTH", away.allele="ANORTH")

#RAT_2011
beta.by.bin.plots(up.exp="RAT_2011", comp.exp="RAM_2011", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="RAT_2011", comp.exp="ADA_2011", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="RAT_2011", comp.exp="ULL_2011", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="RAT_2011", comp.exp="RAT_2012", home.allele="ASOUTH", away.allele="ANORTH")

#ULL_2012
beta.by.bin.plots(up.exp="RAT_2012", comp.exp="RAM_2012", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="RAT_2012", comp.exp="ADA_2012", home.allele="ASOUTH", away.allele="ANORTH")
beta.by.bin.plots(up.exp="RAT_2012", comp.exp="ULL_2012", home.allele="ASOUTH", away.allele="ANORTH")

#ADA_2011
beta.by.bin.plots(up.exp="ADA_2011", comp.exp="RAM_2011", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="ADA_2011", comp.exp="RAT_2011", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="ADA_2011", comp.exp="ULL_2011", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="ADA_2011", comp.exp="ADA_2012", home.allele="ANORTH", away.allele="ASOUTH")


#ADA_2012
beta.by.bin.plots(up.exp="ADA_2012", comp.exp="RAM_2012", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="ADA_2012", comp.exp="RAT_2012", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="ADA_2012", comp.exp="ULL_2012", home.allele="ANORTH", away.allele="ASOUTH")

#RAM_2011
beta.by.bin.plots(up.exp="RAM_2011", comp.exp="ADA_2011", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="RAM_2011", comp.exp="RAT_2011", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="RAM_2011", comp.exp="ULL_2011", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="RAM_2011", comp.exp="RAM_2012", home.allele="ANORTH", away.allele="ASOUTH")


#RAM_2012
beta.by.bin.plots(up.exp="RAM_2012", comp.exp="ADA_2012", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="RAM_2012", comp.exp="RAT_2012", home.allele="ANORTH", away.allele="ASOUTH")
beta.by.bin.plots(up.exp="RAM_2012", comp.exp="ULL_2012", home.allele="ANORTH", away.allele="ASOUTH")


beta.by.bin.plots(up.exp="ULL_2011", comp.exp="ULL_2012", home.allele="AS1", away.allele="AS2")
beta.by.bin.plots(up.exp="RAT_2011", comp.exp="RAT_2012", home.allele="AS1", away.allele="AS2")
beta.by.bin.plots(up.exp="ULL_2011", comp.exp="ULL_2012", home.allele="AS2", away.allele="AS1")
beta.by.bin.plots(up.exp="RAT_2011", comp.exp="RAT_2012", home.allele="AS2", away.allele="AS1")

beta.by.bin.plots(up.exp="ULL_2012", comp.exp="RAT_2012", home.allele="AS2", away.allele="AS1")
beta.by.bin.plots(up.exp="ULL_2012", comp.exp="RAT_2012", home.allele="AS1", away.allele="AS2")
beta.by.bin.plots(up.exp="RAT_2012", comp.exp="ULL_2012", home.allele="AS2", away.allele="AS1")
beta.by.bin.plots(up.exp="RAT_2012", comp.exp="ULL_2012", home.allele="AS1", away.allele="AS2")

```


### 11. Manhattan plots of AFD home/away, SNPs colored by home beta
Is there a genome pattern of these?

```{r afd manhattan function}
home.allele = "ANORTH"
away.allele = "ASOUTH"
up.exp = "RAM_2011"
afb.dat<- afb.dat[order(afb.dat$chr, afb.dat$pos),]

afd.manhattan.plot <- function(home.allele, away.allele, up.exp){
  ### polarize by home allele and get AFD home-away
  home.c <- which(colnames(afb.dat)==home.allele)
  away.c <- which(colnames(afb.dat)==away.allele)
  maf.dat <- afb.dat[afb.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat[17:24] <- -maf.dat[17:24]
  
  # do same to subset where alt allele is major allele in home pop (don't need to flip sign)
  home.dat <- afb.dat[afb.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  
  ### generate plot
  plot.title <- up.exp
  plot.subtitle <- paste("home=",home.allele, " away=",away.allele, sep="")
  man.plot <- ggplot(out.dat, aes(x=pos/1000000, y=get(up.exp), color=ha.bins)) +
    geom_point() +
    facet_grid(~chrom) +
    #scale_color_viridis(direction=-1) +
    scale_fill_viridis(direction=-1) +
    theme_bw() +
    xlab("chromosome and position (Mb)") + 
    ylab(paste("home allele effect in ",up.exp, sep="")) + 
    ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), ""))))+
    geom_hline(yintercept = 0, color="red", linetype="dashed") +
    labs(color = "AFD bin")
  
  plot.file <- paste("./figures/47.figures/afd.beta.manhattan",up.exp,home.allele,away.allele,"jpg",sep=".")
  ggsave(file=plot.file, man.plot, width=14, height=4, units="in")
}
```

```{r afd manhattan plots}

## north plots
for(up.e in c("RAM_2011","RAM_2012","ADA_2011","ADA_2012")){
  print(up.e)
  afd.manhattan.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp=up.e)
}

for(up.e in c("RAM_2011","RAM_2012","ADA_2011","ADA_2012")){
  print(up.e)
  afd.manhattan.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp=up.e)
}


## south plots
for(up.e in c("RAT_2011","RAT_2012","ULL_2011","ULL_2012")){
  print(up.e)
  afd.manhattan.plot(home.allele="ASOUTH", away.allele="ANORTH", up.exp=up.e)
}

for(up.e in c("RAT_2011","RAT_2012","ULL_2011","ULL_2012")){
  print(up.e)
  afd.manhattan.plot(home.allele="ANORTH", away.allele="ASOUTH", up.exp=up.e)
  }

for(up.e in c("RAT_2011","RAT_2012","ULL_2011","ULL_2012")){
  print(up.e)
  afd.manhattan.plot(home.allele="AS1", away.allele="AS2", up.exp=up.e)
  }

for(up.e in c("RAT_2011","RAT_2012","ULL_2011","ULL_2012")){
  print(up.e)
  afd.manhattan.plot(home.allele="AS2", away.allele="AS1", up.exp=up.e)
  }

```










### appendix: Figures for SAB report
```{r figures for SAB}

### RAM_2011 boxplots by AFD
up.f <- "./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds"
up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
up.pheno <- get.p.name(up.gwa.file=up.f)
nsa.plot <- ggplot(data = up.dat, mapping = aes(x=ha.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='allele frequency difference North-South populations (polarized by most common allele in North)', y="effect of most common allele in North") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal()
 nsa.plot <- nsa.plot + ggtitle("experiment NR 2011",subtitle="home allele=NORTH, away allele=SOUTH")
 plot.name <- "./figures/47.figures/beta.afd.boxplots.NR_2011.forSAB.jpg"
 ggsave(plot.name, plot=nsa.plot, width=8, height=6, units="in")

### ULL_2011 boxplots by AFD
up.f <- "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds"
up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
up.pheno <- get.p.name(up.gwa.file=up.f)
sna.plot <- ggplot(data = up.dat, mapping = aes(x=ha.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='allele frequency difference South-North populations (polarized by most common allele in South)', y="effect of most common allele in South") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal()
 sna.plot <- sna.plot + ggtitle("experiment SU 2011",subtitle="home allele=SOUTH, away allele=NORTH")
 plot.name <- "./figures/47.figures/beta.afd.boxplots.SU_2011.forSAB.jpg"
 ggsave(plot.name, plot=sna.plot, width=8, height=6, units="in")

 ### put both of these together
plot.file <-"./figures/47.figures/ULL_2011andRAM_2011.combined.afd.boxplots.forSAB.jpg"
comp.plot <- plot_grid(nsa.plot, sna.plot, labels = c('A', 'B'), ncol = 1, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=8, height=8, units="in")
 
 
### betas pairwise ULL_2011
up.exp <- "ULL_2011"
comp.exp <- "RAM_2011"
home.allele <- "ASOUTH"
away.allele <- "ANORTH"
  ## essentially, repolarizing alleles so away allele effect=0 and data file reports home allele effect
  # get subset where alt allele freq < 0.5 in home population
  home.c <- which(colnames(afb.dat)==home.allele)
  away.c <- which(colnames(afb.dat)==away.allele)
  maf.dat <- afb.dat[afb.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat[17:24] <- -maf.dat[17:24]
  
  # do same to subset where alt allele is major allele in home pop (don't need to flip sign)
  home.dat <- afb.dat[afb.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  
  #do home allele frequency bins
  breaks <- seq(0.5,1,0.1)
  tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$ahome.bins <- cut(out.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ahome.bins <- factor(out.dat$ahome.bins, levels = tags,ordered = TRUE)
  
  #do away allele frequency bins
  breaks <- seq(0,1,0.1)
  tags <- c("[.0-.1)","[.1-.2)","[.2-.3)","[.3-.4)","[.4-5)","[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$aaway.bins <- cut(out.dat$aaway, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$aaway.bins <- factor(out.dat$aaway.bins, levels = tags,ordered = TRUE)
  
  plot.title <- "SU_2011 - NR_2011"
  plot.subtitle <- "home allele=SOUTH, away allele=NORTH"

c.fit <- matrix(nrow=length(levels(out.dat$ha.bins)), ncol=4)
for(up in 1:length(levels(out.dat$ha.bins))){
  up.b <- levels(out.dat$ha.bins)[up]
  up.dat <- out.dat[out.dat$ha.bins==up.b,]
  ue.col <- which(colnames(up.dat)==up.exp)
  ce.col <- which(colnames(up.dat)==comp.exp)
  sh1 <- sum(up.dat$ahome*up.dat[,ue.col], na.rm=TRUE)
  sa1 <- sum(up.dat$aaway*up.dat[,ue.col], na.rm=TRUE)
  sh2 <- sum(up.dat$ahome*up.dat[,ce.col], na.rm=TRUE)
  sa2 <- sum(up.dat$aaway*up.dat[,ce.col], na.rm=TRUE)
  out.fit <- c(sh1, sa1 ,sh2,sa2)
  c.fit[up,] <-out.fit
}
colnames(c.fit) <- c("sh1", "sa1","sh2","sa2")
rownames(c.fit) <- levels(out.dat$ha.bins)
c.fit <- as.data.frame(c.fit)
c.fit$bin <- rownames(c.fit)
c.fit.l <- gather(c.fit, pop.exp, c.fitness, sh1, sa1, sh2, sa2)
c.fit.l$pop.exp <- factor(c.fit.l$pop.exp, levels=c("sh1","sa1","sh2","sa2"))
c.fit.l$bin <- factor(c.fit.l$bin, levels=c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)"))

#number of SNPs in each bin
hen <- out.dat[is.na(out.dat[,ue.col])==FALSE,]
ht <- table(hen$ha.bins)
ht <- as.matrix(ht)
aen <- out.dat[is.na(out.dat[,ce.col])==FALSE,]
at <- table(aen$ha.bins)
at <- as.matrix(at)
sn.index <- cbind(ht,ht,at,at)
colnames(sn.index) <- c("sh1","sa1","sh2","sa2")
sn.index <- as.data.frame(sn.index)
sn.index$bin <- rownames(sn.index)
sn.index.l <- gather(sn.index, pop.exp, snp.no, sh1, sa1, sh2, sa2)

c.fit.l <- merge(c.fit.l, sn.index.l, by=c("pop.exp", "bin"))
c.fit.l$fit.per.snp <- c.fit.l$c.fitness/c.fit.l$snp.no


f.plot <-ggplot(c.fit.l, aes(x=bin, y=c.fitness, group=pop.exp)) + 
  geom_line(aes(color=pop.exp)) + 
  geom_point(aes(color=pop.exp, size=abs(fit.per.snp))) + 
  xlab("allele frequency difference bin (south-north)") + 
  ylab("population fitness of bin") + 
  #scale_color_discrete(name = "population/EXPERIMENT", labels = c("south/SU", "north/SU", "south/NR","north/NR")) +
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), "")))) +
  geom_hline(yintercept = 0, color="grey50", linetype="dashed") +
  scale_color_manual(values=c("red", "pink2", "dodgerblue","lightblue2"),name = "EXPERIMENT/population", labels = c("SU/south", "SU/north", "NR/south","NR/north")) +
  scale_size(name="mean effect per SNP\n(absolute value)") +
  theme_linedraw()  

pw.bins.plot <- ggplot(data=out.dat, aes(x=get(up.exp), y=get(comp.exp))) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  facet_grid(~ha.bins) +
  geom_vline(xintercept = 0, color="orange") +
  geom_hline(yintercept = 0, color="orange") +
  xlab("allele effect SU_2011") +
  ylab("allele effect NR_2011")+
  theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
  stat_quadrant_counts(aes(label = stat(count))) +
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), "")))) 

#generate plots
plot.file <-"./figures/47.figures/ULL_2011.RAM_2011_beta_pairwise_by_AFD_bins.forSAB.pdf"
#pdf(file=plot.file, width=14, height=8)
comp.plot <- plot_grid(pw.bins.plot, labels = c('A', 'B'),f.plot, ncol = 1, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=14, height=8, units="in")
#dev.off()
```


```{r SAB plot part 2}
### betas pairwise RAM_2011
up.exp <- "RAM_2011"
comp.exp <- "ULL_2011"
home.allele <- "ANORTH"
away.allele <- "ASOUTH"
  ## essentially, repolarizing alleles so away allele effect=0 and data file reports home allele effect
  # get subset where alt allele freq < 0.5 in home population
  home.c <- which(colnames(afb.dat)==home.allele)
  away.c <- which(colnames(afb.dat)==away.allele)
  maf.dat <- afb.dat[afb.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat[17:24] <- -maf.dat[17:24]
  
  # do same to subset where alt allele is major allele in home pop (don't need to flip sign)
  home.dat <- afb.dat[afb.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  
  #do home allele frequency bins
  breaks <- seq(0.5,1,0.1)
  tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$ahome.bins <- cut(out.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ahome.bins <- factor(out.dat$ahome.bins, levels = tags,ordered = TRUE)
  
  #do away allele frequency bins
  breaks <- seq(0,1,0.1)
  tags <- c("[.0-.1)","[.1-.2)","[.2-.3)","[.3-.4)","[.4-5)","[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )
  out.dat$aaway.bins <- cut(out.dat$aaway, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$aaway.bins <- factor(out.dat$aaway.bins, levels = tags,ordered = TRUE)
  
  plot.title <- "NR_2011 - SU_2011"
  plot.subtitle <- "home allele=NORTH, away allele=SOUTH"

c.fit <- matrix(nrow=length(levels(out.dat$ha.bins)), ncol=4)
for(up in 1:length(levels(out.dat$ha.bins))){
  up.b <- levels(out.dat$ha.bins)[up]
  up.dat <- out.dat[out.dat$ha.bins==up.b,]
  ue.col <- which(colnames(up.dat)==up.exp)
  ce.col <- which(colnames(up.dat)==comp.exp)
  sh1 <- sum(up.dat$ahome*up.dat[,ue.col], na.rm=TRUE)
  sa1 <- sum(up.dat$aaway*up.dat[,ue.col], na.rm=TRUE)
  sh2 <- sum(up.dat$ahome*up.dat[,ce.col], na.rm=TRUE)
  sa2 <- sum(up.dat$aaway*up.dat[,ce.col], na.rm=TRUE)
  out.fit <- c(sh1, sa1 ,sh2,sa2)
  c.fit[up,] <-out.fit
}
colnames(c.fit) <- c("sh1", "sa1","sh2","sa2")
rownames(c.fit) <- levels(out.dat$ha.bins)
c.fit <- as.data.frame(c.fit)
c.fit$bin <- rownames(c.fit)
c.fit.l <- gather(c.fit, pop.exp, c.fitness, sh1, sa1, sh2, sa2)
c.fit.l$pop.exp <- factor(c.fit.l$pop.exp, levels=c("sh1","sa1","sh2","sa2"))
c.fit.l$bin <- factor(c.fit.l$bin, levels=c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)"))

#number of SNPs in each bin
hen <- out.dat[is.na(out.dat[,ue.col])==FALSE,]
ht <- table(hen$ha.bins)
ht <- as.matrix(ht)
aen <- out.dat[is.na(out.dat[,ce.col])==FALSE,]
at <- table(aen$ha.bins)
at <- as.matrix(at)
sn.index <- cbind(ht,ht,at,at)
colnames(sn.index) <- c("sh1","sa1","sh2","sa2")
sn.index <- as.data.frame(sn.index)
sn.index$bin <- rownames(sn.index)
sn.index.l <- gather(sn.index, pop.exp, snp.no, sh1, sa1, sh2, sa2)

c.fit.l <- merge(c.fit.l, sn.index.l, by=c("pop.exp", "bin"))
c.fit.l$fit.per.snp <- c.fit.l$c.fitness/c.fit.l$snp.no


f.plot <-ggplot(c.fit.l, aes(x=bin, y=c.fitness, group=pop.exp)) + 
  geom_line(aes(color=pop.exp)) + 
  geom_point(aes(color=pop.exp, size=abs(fit.per.snp))) + 
  xlab("allele frequency difference bin (north-south)") + 
  ylab("population fitness of bin") + 
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), "")))) +
  geom_hline(yintercept = 0, color="grey50", linetype="dashed") +
  scale_color_manual(values=c("red", "pink2", "dodgerblue","lightblue2"),name = "EXPERIMENT/population", labels = c("NR/north", "NR/south", "SU/north","SU/south")) +
  scale_size(name="mean effect per SNP\n(absolute value)") +
  theme_linedraw()  

pw.bins.plot <- ggplot(data=out.dat, aes(x=get(up.exp), y=get(comp.exp))) +
  geom_bin2d(bins = 100) +
  scale_fill_continuous(type = "viridis") +
  facet_grid(~ha.bins) +
  geom_vline(xintercept = 0, color="orange") +
  geom_hline(yintercept = 0, color="orange") +
  xlab("allele effect NR_2011") +
  ylab("allele effect SU_2011")+
  theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
  stat_quadrant_counts(aes(label = stat(count))) +
  ggtitle(bquote(atop(.(plot.title), atop(italic(.(plot.subtitle)), "")))) 

#generate plots
plot.file <-"./figures/47.figures/RAM_2011.ULL_2011_beta_pairwise_by_AFD_bins.forSAB.pdf"
#pdf(file=plot.file, width=14, height=8)
comp.plot <- plot_grid(pw.bins.plot, labels = c('A', 'B'),f.plot, ncol = 1, align="v", axis="rl")
ggsave(file=plot.file, comp.plot, width=14, height=8, units="in")
#dev.off()





```

This section is a mess of strange ideas.  Not run.  move to end so can keep script in case needed???
```{r median of co- vs opposite effects}
## calculating the centroid along the x/y axes does tell us something, but is it the right thing?
## i propose rotating everything 45 degrees, then doing the centroid calculations.  This will tell us
## the relative balance between opposing effects and between common effects!!!
## angle is radians counterclockwise
rotate.xy <- function (xy, angle) {
    xy <- as.matrix(xy)
    cos.angle <- cos(angle)
    sin.angle <- sin(angle)
    tx <- xy[,1]*cos.angle - xy[,2]*sin.angle
    ty <- xy[,1]*sin.angle + xy[,2]*cos.angle
    tout <- cbind(tx,ty)
    return(tout)
}

h.exp <- "ULL_2011"
a.exp <- "RAM_2011"
up.afg <- afb.s[afb.s$ha.bins=="[.85-1)",]
up.coords <- up.afg[,c(h.exp, a.exp)]

colnames(up.coords.45) <- c("common", "opposite")
up.coords.45 <- rotate.xy(up.coords,-45*pi/180)
up.coords.45 <- as.data.frame(up.coords.45)
cmean <- mean(up.coords.45[,1], na.rm=TRUE)
omean <- mean(up.coords.45[,2], na.rm=TRUE)
mean.45 <- matrix(c(cmean, 0, 0, omean), ncol=2)
mean.oc <- rotate.xy(mean.45, 45*pi/180)



up.dat <- afb.s

#generate centroids per bin
maxb <- max(abs(up.dat[,colnames(up.dat)%in%c(h.exp, a.exp)]),na.rm=TRUE)
hmean <- aggregate(get(h.exp)~ha.bins, data=up.dat, mean, na.action = na.omit)
amean <- aggregate(get(a.exp)~ha.bins, data=up.dat, mean, na.action = na.omit)
bmeans <- merge(hmean, amean)
colnames(bmeans)[2:3] <- c("h.mean", "a.mean")

ggplot(data=afb.s, aes(x=get(h.exp), y=get(a.exp))) +
    geom_density_2d() +
    facet_grid(~ ha.bins) +
    geom_vline(xintercept = 0, color="orange") +
    geom_hline(yintercept = 0, color="orange") +
    xlab(h.exp) +
    ylab(a.exp)+
    theme(axis.text.x=element_text(angle =- 90, vjust = 0.5)) +
    stat_quadrant_counts(aes(label = stat(count))) +
    geom_point(data=bmeans, aes(x=h.mean, y=a.mean), color="green")

+
    #geom_segment(aes(x = 0, y = 0, xend = mean.oc[1,1], yend = mean.oc[1,2]),arrow = arrow(length = unit(0.25, "cm")), color="red") +
    #geom_segment(aes(x = 0, y = 0, xend = mean.oc[2,1], yend = mean.oc[2,2]),arrow = arrow(length = unit(0.25, "cm")), color="green") +
    xlim(-maxb, maxb) +
    ylim(-maxb, maxb) 



+
    geom_point(x=mean(up.afg[,h.exp]), y=mean(up.afg[,a.exp]), color="green")


    geom_point(aes(x=mean.oc[1], y=mean.oc[2]), colour="green", size=2)
    
    
```



