---
title: "48.AFD.genome.scan.subsets.Rmd"
author: "Daniele Filiault"
date: "15MAR2021"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Volumes/field_experiments/adaptation_sweden/common.gardens/')

library(ggpmisc)
library(ggplot2)
library(gridExtra)
library(dplyr)
library(tidyr)
library(ggpubr)
library(cowplot)
library(viridis)
library(rcompanion)
library(multcompView)
library(egg)
#library(spdep)
source("99.geo.helper.fxns.R") ### has my genome plot function.

#setwd("/groups/nordborg/projects/field_experiments/adaptation_sweden/common.gardens")
setwd("/Volumes/field_experiments/adaptation_sweden/common.gardens/")
```

## Introduction

Look at patterns of allele frequency vs beta for subsets of SNPs identified in genome scans to see if relationships are same or different from genome-wide. 

## 1. Load AFD data

```{r load data}
### allele frequency data from script 46.allele.freq.differences.Kgroups
load("./data/pop.af.dat.Rdat")
load("./data/allele.freq.GWAS.snps.Rdata") #afg
```

```{r prep AFD data}
### combine AFD datasets
colnames(pop.af.dat)[1:2] <- c("chrom", "pos")
pos <- do.call(rbind,strsplit(rownames(afg),"_"))
colnames(pos) <- c("chrom", "pos")
pos <- as.data.frame(pos)
afg <- cbind(afg, pos)
af.dat <- merge(pop.af.dat, afg, by=c("chrom", "pos"), all=TRUE)
```

## 2. Add GWAS data and polarize by most frequent local allele

Data ready to go.  Now for each GWAS result, need to load and prep data, then make boxplots.  Write functions for these

```{r data prep fxns}

### function to combine GWAS and AFD data and polarize by "home" and "away" alleles

add.gwas <- function(up.gwa.file, home.allele, away.allele){
  load(up.gwa.file) ##variable is named p
  colnames(p)[1:3] <- c("chrom","marker","pos")
  up.dat <- merge(af.dat, p, by=c("chrom", "pos"),all=TRUE)
  
  # get subset where alt allele freq < 0.5 in home population
  home.c <- which(colnames(up.dat)==home.allele)
  away.c <- which(colnames(up.dat)==away.allele)
  maf.dat <- up.dat[up.dat[,home.c]<0.5,]
  #these are SNPs where the ref allele is more frequent locally
  #need to calculate AFD and switch beta appropriately
  maf.dat$ahome <- 1-maf.dat[,home.c]
  maf.dat$aaway <- 1-maf.dat[,away.c]
  maf.dat$home.beta <- -(maf.dat$beta)
  
  # do same to subset where alt allele is major allele in home pop
  home.dat <- up.dat[up.dat[,home.c]>=0.5,]
  home.dat$ahome <- home.dat[,home.c]
  home.dat$aaway <- home.dat[,away.c]
  home.dat$home.beta <- home.dat$beta 
  
  # put back together and finish AFD calculations
  out.dat <- rbind(maf.dat, home.dat)
  out.dat$ha.afd <- out.dat$ahome-out.dat$aaway
  # get AFD bins
  breaks <- seq(-0.5,1,0.15)
  # specify interval/bin labels
  tags <- c("-[.5-.35)","-[.35-.2)", "-[.2-.05)", "-[.05-.1)", "[-.1-.25)","[.25-.4)", "[.4-.55)","[.55-.7)", "[.7-85)","[.85-1)")
  # bucketing values into bins north/south
  out.dat$ha.bins <- cut(out.dat$ha.afd, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
  out.dat$ha.bins <- factor(out.dat$ha.bins, levels = tags,ordered = TRUE)
  return(out.dat)
}

### extract pheno name from filename
get.p.name <- function(up.gwa.file){
  up.pheno <- unlist(strsplit(up.gwa.file, "/"))
  up.pheno <- up.pheno[grep("rds", up.pheno)]
  up.pheno <- gsub("gemma_lmm_blup_","" ,up.pheno)
  up.pheno <- gsub(".rds","" ,up.pheno)
  return(up.pheno)
}

### make pheno names to short names
short.p.name <- function(p.name){
  p.name <- gsub("ADA","NA",p.name)
  p.name <- gsub("RAM","NM",p.name)
  p.name <- gsub("RAT","SR",p.name)
  p.name <- gsub("ULL","SU",p.name)
  return(p.name)
}
```


```{r polarizebetas}
gwas.res.files.n <- c("./res/gemma_marginal/gemma_lmm_blup_RAM_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAM_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ADA_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ADA_2012.rds")
gwas.res.files.s <- c("./res/gemma_marginal/gemma_lmm_blup_RAT_2011.rds", "./res/gemma_marginal/gemma_lmm_blup_RAT_2012.rds", "./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds","./res/gemma_marginal/gemma_lmm_blup_ULL_2012.rds")

home.beta.n <- af.dat[,1:2]
for(up.f in gwas.res.files.n){
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ANORTH", away.allele="ASOUTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  colnames(up.dat)[colnames(up.dat)=="home.beta"] <- up.pheno
  home.beta.n <- merge(home.beta.n, up.dat[,colnames(up.dat)%in%c("chrom", "pos",up.pheno)], all=TRUE)
}

home.beta.s <- af.dat[,1:2]
for(up.f in gwas.res.files.s){
  print(up.f)    
  up.dat <- add.gwas(up.gwa.file=up.f, home.allele="ASOUTH", away.allele="ANORTH")
  up.pheno <- get.p.name(up.gwa.file=up.f)
  colnames(up.dat)[colnames(up.dat)=="home.beta"] <- up.pheno
  home.beta.s <- merge(home.beta.s, up.dat[,colnames(up.dat)%in%c("chrom", "pos",up.pheno)], all=TRUE)
}

home.beta <- merge(home.beta.n, home.beta.s)
home.beta <- merge(home.beta,af.dat[,c(1,2,15:16)])

```

# Developing functions 

## 3. Subset by genome scan data
```{r filter by scan data function}
#up.scan.file <- "./data/003.selection.scans/Fournier_Level_GWAs_Clim_Data.csv"
#up.scan.name <- "fournierlevel"
#win.size <- 10000

#fl.dat <- read.csv(up.scan.file, stringsAsFactors=FALSE)
#ss.dat <- fl.dat[,1:2]
#colnames(ss.dat) <- c("chr", "pos")
## data frame of 2 columns, chr and pos
## might need to do different manipulations for different scans to get here.

snp.in.scan <- function(ss.dat, win.size){
  ## add column to home.beta that says whether a SNP is in the dataset or not
  ss.dat$start <- ss.dat$pos-win.size
  ss.dat$stop <- ss.dat$pos+win.size

  ss.pos <- apply(ss.dat,1, function(x){
    out.dat <- home.beta[home.beta$chrom==x[1] & home.beta$pos>=x[3] & home.beta$pos<=x[4],1:2]
  })
  ss.pos <- do.call(rbind, ss.pos)
  ss.pos <- unique(ss.pos) #68750 positions
  ss.pos$scan <- TRUE

  home.beta.ss <- merge(home.beta, ss.pos, all.x=TRUE)
  scan.beta <- home.beta.ss[!is.na(home.beta.ss$scan),]
  return(scan.beta)
}
## there is now a column called "scan" that ==TRUE if overlapping genome scan

#scan.beta <- snp.in.scan(ss.dat=ss.dat, win.size=10000)
```


## 4. Test effect of home allele

```{r home allele greater than zero function}
test.home.allele.positive <- function(scan.beta){
  #one-sided t-test mean beta >0
  scan.ttest <- apply(scan.beta[,3:10],2,function(x){t.test(x, mu = 0, alternative = "greater")})
  scan.pval <- lapply(scan.ttest,function(x){x$p.value})
  print(scan.pval) #RAM_2011, ADA_2011 and all South are significantly >0.  So for all "local adaptation" (plus RAT_2011), home allele on average has positive fitness effect.
  scan.pval <- unlist(scan.pval)
  scan.pval <- scan.pval[match(c("ADA_2011", "ADA_2012", "RAM_2011", "RAM_2012", "ULL_2011", "ULL_2012", "RAT_2011", "RAT_2012"),names(scan.pval))]
  scan.pval <- format(scan.pval, digits=2)
  return(scan.pval)
}

#fl.pvals <- test.home.allele.positive(scan.beta=scan.beta)
```

```{r home allele nonzero function}
test.home.allele.zero <- function(scan.beta){
  ##two-sided t-test mean beta
  scan.2ttest <- apply(scan.beta[,3:10],2,function(x){t.test(x, mu = 0, alternative = "two.sided")})
  scan.2pval <- lapply(scan.2ttest,function(x){x$p.value})
  print(scan.2pval) #all are sig different from zero, though (ADA_2011 is 0.035)
  scan.2pval <- scan.2pval[match(c("ADA_2011", "ADA_2012", "RAM_2011", "RAM_2012", "ULL_2011", "ULL_2012", "RAT_2011", "RAT_2012"),names(scan.2pval))]
  scan.2pval <- format(scan.2pval, digits=2)
  return(scan.2pval)
}

#fl.nonzero.pvals <- test.home.allele.zero(scan.beta=scan.beta)
```

```{r boxplot beta home allele function}
## boxplot of beta of home allele values

home.allele.boxplot <- function(scan.beta, scan.pval, up.scan.name, win.size){
  #reshape data
  scan.beta.l <- reshape(scan.beta,varying=c(3:10), direction="long", timevar="exp",times=colnames(scan.beta)[3:10],v.name="beta")
  scan.beta.l$exp <- gsub("RAM","NM", scan.beta.l$exp)
  scan.beta.l$exp <- gsub("ADA","NA", scan.beta.l$exp)
  scan.beta.l$exp <- gsub("RAT","SR", scan.beta.l$exp)
  scan.beta.l$exp <- gsub("ULL","SU", scan.beta.l$exp)
  scan.beta.l$exp <- factor(scan.beta.l$exp,levels=c("NA_2011","NA_2012","NM_2011","NM_2012","SU_2011","SU_2012","SR_2011","SR_2012"))
  scan.beta.l$region <- substr(scan.beta.l$exp,1,1)

  home.beta.plot <- ggplot(data = scan.beta.l, mapping = aes(x=exp,y=beta,fill=region)) + 
      #geom_jitter(aes(color='blueviolet'),alpha=0.2) +
      #scale_color_manual(values=c("blueviolet")) +
      #geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
      geom_violin() +
      scale_fill_manual(values=c("dodgerblue1","springgreen2"), name = "experiment \nregion", labels = c("North", "South")) +
      geom_boxplot(width=0.1, fill="white") +
      labs(x='experiment', y="fitness effect of most common allele in home pop") +
      guides(color=FALSE) +
      geom_hline(yintercept=0, linetype="dashed", color = "red", size=0.5) +
      theme_minimal() +
      stat_summary(geom = 'text', label = scan.pval, fun = max, vjust = -1)

  plot.name <- paste("./figures/48.figures/beta.home.allele.violin.plots",up.scan.name,win.size,"jpg",sep=".")
  ggsave(plot.name, plot=home.beta.plot, width=8, height=6, units="in")
}

#home.allele.boxplot(scan.beta=scan.beta, scan.pval=fl.pvals, up.scan.name="fournierlevel", win.size=10000)
```

## 5. Relationship between home allele frequency and beta

```{r test home allele freq vs beta function}
#home.allele="ASOUTH"
#away.allele="ANORTH"
#gwas.files=gwas.res.files.s

haf.beta.boxplot <- function(home.allele, away.allele, gwas.files, scan.beta){
  s.ahome.plots <- as.list(1:4)
  s.mw.test <- as.list(1:4)
  for(up.fn in 1:4){
    up.f <- gwas.files[up.fn]
    print(up.f)    
    up.dat <- add.gwas(up.gwa.file=up.f, home.allele=home.allele, away.allele=away.allele)
    #subset for selection scan positions
    up.dat <- merge(up.dat,scan.beta[,1:2])
    up.pheno <- get.p.name(up.gwa.file=up.f)
    up.short.name <- short.p.name(p.name=up.pheno)
    # get AFD bins
    breaks <- seq(0.5,1,0.1)
    # specify interval/bin labels
    tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )  # bucketing values into bins north/south
    up.dat$ahome.bins <- cut(up.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
    up.dat$ahome.bins <- factor(up.dat$ahome.bins, levels = tags,ordered = TRUE)
  
    #test difs between allele freq bins
    pt <- pairwise.wilcox.test(up.dat$home.beta, up.dat$ahome.bins, p.adj.method="bh")
    PT <- pt$p.value
    library(rcompanion)
    PT1 = fullPTable(PT)
    comp.lets <- multcompLetters(PT1,compare="<",threshold=0.05,Letters=letters,reversed = FALSE)
    s.mw.test[[up.fn]] <- pt

    haf.boxplot <- ggplot(data = subset(up.dat, !is.na(ahome.bins)), mapping = aes(x=ahome.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='home allele frequency', y="home allele effect") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal() +
    ggtitle(bquote(atop(.(up.short.name), atop(italic(.("home allele = SOUTH")), "")))) +
    stat_summary(geom = 'text', label = comp.lets$Letters, fun = max, vjust = -1)

    s.ahome.plots[[up.fn]] <- haf.boxplot
    }
  return(s.ahome.plots)
  }

#south.plots <- haf.beta.boxplot(home.allele="ASOUTH", away.allele="ANORTH", gwas.files=gwas.res.files.s, scan.beta=scan.beta)

#north.plots <- haf.beta.boxplot(home.allele="ANORTH", away.allele="ASOUTH", gwas.files=gwas.res.files.n, scan.beta=scan.beta)
```

```{r plot home AF vs beta - both sets}
plot.afd.beta <- function(s.ahome.plots, n.ahome.plots, up.scan.name, win.size){
  ## plot all in one file
  plot.file <- paste("./figures/48.figures/home.allele.freq.vs.effect.all.experiments",up.scan.name, win.size,"jpg", sep=".")
  comp.plot <- plot_grid(n.ahome.plots[[3]], n.ahome.plots[[1]], s.ahome.plots[[3]], s.ahome.plots[[1]], n.ahome.plots[[4]], n.ahome.plots[[2]], s.ahome.plots[[4]], s.ahome.plots[[2]], ncol = 4, align="v", axis="rl")
  ggsave(file=plot.file, comp.plot, width=15, height=12, units="in")
}

#plot.afd.beta(s.ahome.plots=south.plots, n.ahome.plots=north.plots, up.scan.name="fournierlevel",win.size=10000)
```

```{r scan and nonscan AF plot and test}
## coplot scan and non-scan data and test for differences within each AF group
#scan.beta <- snp.in.scan(ss.dat=sg.dat, win.size=win.size)

#home.allele <- "ANORTH"
#away.allele <- "ASOUTH"
#gwas.files=gwas.res.files.n
#scan.beta=scan.beta
#n.comp.af.boxplots <- test.af.boxplot(home.allele=home.allele, away.allele=away.allele, gwas.files=gwas.files, scan.beta=scan.beta)

#home.allele <- "ASOUTH"
#away.allele <- "ANORTH"
#gwas.files=gwas.res.files.s
#scan.beta=scan.beta
#s.comp.af.boxplots <- test.af.boxplot(home.allele=home.allele, away.allele=away.allele, gwas.files=gwas.files, scan.beta=scan.beta)

test.af.boxplot <- function(home.allele, away.allele, gwas.files, scan.beta){
  ahome.plots <- as.list(1:4)
  for(up.fn in 1:4){
    up.f <- gwas.files[up.fn]
    print(up.f)    
    up.dat <- add.gwas(up.gwa.file=up.f, home.allele=home.allele, away.allele=away.allele)
    #subset for selection scan positions
    up.dat <- merge(up.dat, scan.beta[,c(1:2,13)], all.x=TRUE)
    up.dat$in.ss <- (is.na(up.dat$scan)==FALSE)
    up.dat$in.ss <- factor(up.dat$in.ss, levels=c("TRUE", "FALSE"))
    up.pheno <- get.p.name(up.gwa.file=up.f)
    up.short.name <- short.p.name(p.name=up.pheno)
    # get AFD bins
    breaks <- seq(0.5,1,0.1)
    # specify interval/bin labels
    tags <- c("[.5-.6)","[.6-.7)","[.7-.8)","[.8-.9)","[.9-1)" )  # bucketing values into bins north/south
    up.dat$ahome.bins <- cut(up.dat$ahome, breaks=breaks, include.lowest=TRUE, right=FALSE, labels=tags)
    up.dat$ahome.bins <- factor(up.dat$ahome.bins, levels = tags,ordered = TRUE)
    
    ## pairwise t-test between scan and non-scan data per AFD bin
    t.dat <- split(up.dat, up.dat$ahome.bins)
    t.pvals <- lapply(t.dat, function(x.dat){
      ttx <- t.test(home.beta~in.ss, data=x.dat)
      return(ttx$p.value)
    })
    t.pvals <- do.call(rbind,t.pvals)
    t.pvals <- format(t.pvals, digits=2)
    
    haf.boxplot <- ggplot(data = up.dat, mapping = aes(x=in.ss,y=home.beta, fill=in.ss)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(color="black", alpha=0.7) + 
    labs(x='home allele frequency', y="home allele effect") +
    facet_wrap(.~ahome.bins, nrow=1,strip.position = "bottom") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal() +
    theme(axis.text.x=element_blank())+
    ggtitle(up.short.name) +
    scale_fill_discrete(name="selection\nscan")
    
    ymax <- ggplot_build(haf.boxplot)$layout$panel_scales_y[[1]]$range$range[2]
    xpos <- mean(ggplot_build(haf.boxplot)$layout$panel_scales_x[[1]]$range_c$range)
    haf.boxplot <- tag_facet(haf.boxplot, x=xpos, y=ymax, open = "", close = "",tag_pool = t.pvals, fontface=1, size=3, hjust=0.5,vjust=-1) +
      theme(strip.text = element_text(),strip.background = element_rect())
  
    ahome.plots[[up.fn]] <- haf.boxplot
    }
  return(ahome.plots)
  }

## plot all in one file
scan.nonscan.af.plot.output <- function(n.comp.af.boxplots, s.comp.af.boxplots, up.scan.name, win.size){
  plot.file <- paste("./figures/48.figures/AFD.scan.nonscan.all.experiments",up.scan.name, win.size,"jpg",sep=".")
  comp.plot <- plot_grid(n.comp.af.boxplots[[3]], n.comp.af.boxplots[[1]], s.comp.af.boxplots[[3]], s.comp.af.boxplots[[1]], n.comp.af.boxplots[[4]], n.comp.af.boxplots[[2]], s.comp.af.boxplots[[4]], s.comp.af.boxplots[[2]], ncol = 4, align="v", axis="rl")
  ggsave(file=plot.file, comp.plot, width=24, height=10, units="in")
}
```

## 6. Barplots by AFD bins (polarized by most frequent "local" i.e. N/S allele)

```{r plot betas by allele frequency bins function}
#up.dat <- add.gwas(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds", home.allele="AS2", away.allele="AS1")
#up.pheno <- get.p.name(up.gwa.file="./res/gemma_marginal/gemma_lmm_blup_ULL_2011.rds")

barplot.home.af.bin <- function(up.dat, up.pheno, home.allele, away.allele, comp.lets){
  nsa.plot <- ggplot(data = up.dat, mapping = aes(x=ha.bins,y=home.beta)) + 
    geom_jitter(aes(color='blueviolet'),alpha=0.2) +
    scale_color_manual(values=c("blueviolet")) +
    geom_boxplot(fill="aquamarine",color="black", alpha=0.7) + 
    labs(x='difference between home and away allele frequency', y="home allele effect") +
    guides(color=FALSE) +
    geom_hline(yintercept=0, linetype="dashed", color = "red", size=1) +
    theme_minimal() +
    stat_summary(geom = 'text', label = comp.lets$Letters, fun = max, vjust = -1)
 nsa.plot <- nsa.plot + ggtitle(up.pheno,subtitle = paste("home allele=",home.allele,", away allele=", away.allele, sep=""))
}

```

```{r plot BLUP GWAS afd results functions}

afd.plot <- function(home.allele, away.allele, gwas.files, scan.beta){
  n.afd.plots <- as.list(1:4)
  n.mw.afd.test <- as.list(1:4)

  for(up.fn in 1:4){
    up.f <- gwas.files[up.fn]
    print(up.f)    
    up.dat <- add.gwas(up.gwa.file=up.f, home.allele=home.allele, away.allele=away.allele)
    #subset for selection scan
    up.dat <- merge(up.dat,scan.beta[,1:2])
    up.pheno <- get.p.name(up.gwa.file=up.f)
    up.short.name <- short.p.name(p.name=up.pheno)
    
    #test difs between allele freq bins
    pt <- pairwise.wilcox.test(up.dat$home.beta, up.dat$ha.bins, p.adj.method="bh")
    PT <- pt$p.value
    PT1 = fullPTable(PT)
    comp.lets <- multcompLetters(PT1,compare="<",threshold=0.05,Letters=letters,reversed = FALSE)
    n.mw.afd.test[[up.fn]] <- pt
  
    out.barplot <- barplot.home.af.bin(up.dat=up.dat, up.pheno=up.short.name, home.allele=home.allele, away.allele=away.allele, comp.lets=comp.lets)
    n.afd.plots[[up.fn]] <- out.barplot
  }
  return(n.afd.plots)
}

#north.afd.plots <- afd.plot(home.allele="ANORTH", away.allele="ASOUTH", gwas.files=gwas.res.files.n, scan.beta=scan.beta)
#south.afd.plots <- afd.plot(home.allele="ASOUTH", away.allele="ANORTH", gwas.file=gwas.res.files.s, scan.beta=scan.beta)

## plot all in one file
afd.plot.output <- function(n.afd.plots, s.afd.plots, up.scan.name, win.size){
  plot.file <- paste("./figures/48.figures/AFD.vs.effect.all.experiments",up.scan.name, win.size,"jpg",sep=".")
  comp.plot <- plot_grid(n.afd.plots[[3]], n.afd.plots[[1]], s.afd.plots[[3]], s.afd.plots[[1]], n.afd.plots[[4]], n.afd.plots[[2]], s.afd.plots[[4]], s.afd.plots[[2]], ncol = 4, align="v", axis="rl")
  ggsave(file=plot.file, comp.plot, width=20, height=12, units="in")
}

#afd.plot.output(n.afd.plots=north.afd.plots, s.afd.plots=south.afd.plots, up.scan.name="fournierlevel", win.size=10000)
```

OK, this has now generated a pipeline of functions so i can do this for whatever selection scan.

# Running analysis for genome scans

## 7. Fournier-Level selection scan
```{r generate analysis fournier level data}
# prep data
up.scan.file <- "./data/003.selection.scans/Fournier_Level_GWAs_Clim_Data.csv"
up.scan.name <- "fournier.level"
win.size <- 10000

fl.dat <- read.csv(up.scan.file, stringsAsFactors=FALSE)
fl.dat <- fl.dat[,1:2]
colnames(fl.dat) <- c("chr", "pos")
### also try subset for FIN only

#### all this should run automatically
#### make it one function?

select.scan.plots <- function(ss.dat, up.scan.name, win.size){
  ## test effect of most common allele
  scan.beta <- snp.in.scan(ss.dat=ss.dat, win.size=win.size)
  fl.pvals <- test.home.allele.positive(scan.beta=scan.beta)
  fl.nonzero.pvals <- test.home.allele.zero(scan.beta=scan.beta)
  home.allele.boxplot(scan.beta=scan.beta, scan.pval=fl.pvals, up.scan.name=up.scan.name, win.size=10000)

  ## effect of most common allele broken down by home frequency
  south.plots <- haf.beta.boxplot(home.allele="ASOUTH", away.allele="ANORTH", gwas.files=gwas.res.files.s, scan.beta=scan.beta)
  north.plots <- haf.beta.boxplot(home.allele="ANORTH", away.allele="ASOUTH", gwas.files=gwas.res.files.n, scan.beta=scan.beta)
  plot.afd.beta(s.ahome.plots=south.plots, n.ahome.plots=north.plots, up.scan.name=up.scan.name,win.size=win.size)
  
  ## effect of most common allele by home frequency - comparing sweep and non-sweep SNPs
  n.comp.af.boxplots <- test.af.boxplot(home.allele="ANORTH", away.allele="ASOUTH", gwas.files=gwas.res.files.n, scan.beta=scan.beta)
  s.comp.af.boxplots <- test.af.boxplot(home.allele="ASOUTH", away.allele="ANORTH", gwas.files=gwas.res.files.s, scan.beta=scan.beta)
  scan.nonscan.af.plot.output(n.comp.af.boxplots=n.comp.af.boxplots, s.comp.af.boxplots=s.comp.af.boxplots, up.scan.name=up.scan.name, win.size=win.size)

  ## effect of most common allele broken down by allele frequency difference
  north.afd.plots <- afd.plot(home.allele="ANORTH", away.allele="ASOUTH", gwas.files=gwas.res.files.n, scan.beta=scan.beta)
  south.afd.plots <- afd.plot(home.allele="ASOUTH", away.allele="ANORTH", gwas.file=gwas.res.files.s, scan.beta=scan.beta)
  afd.plot.output(n.afd.plots=north.afd.plots, s.afd.plots=south.afd.plots, up.scan.name=up.scan.name, win.size=win.size)
}

## effect of most common allele by allele freq difference - comparing sweep and non-sweep SNPs
  n.comp.afd.boxplots <- test.afd.boxplot(home.allele="ANORTH", away.allele="ASOUTH", gwas.files=gwas.res.files.n, scan.beta=scan.beta)
  s.comp.afd.boxplots <- test.afd.boxplot(home.allele="ASOUTH", away.allele="ANORTH", gwas.files=gwas.res.files.s, scan.beta=scan.beta)
  scan.nonscan.afd.plot.output(n.comp.afd.boxplots=n.comp.afd.boxplots, s.comp.afd.boxplots=s.comp.afd.boxplots, up.scan.name=up.scan.name, win.size=win.size)

select.scan.plots(ss.dat=fl.dat, up.scan.name="fournierlevel", win.size=10000)
```

## 8. Swedish genomes paper sweeps

```{r generate analysis swedish genomes sweep data}
# prep data
up.scan.file <- "./data/003.selection.scans/Swedish.genomes.paper.sweeps.simple.csv"
#up.scan.name <- "swedish.genomes"
#win.size <- 10000
sg.dat <- read.csv(up.scan.file, stringsAsFactors=FALSE)
sg.dat <- sg.dat[,1:2]
colnames(sg.dat) <- c("chr", "pos")

select.scan.plots(ss.dat=sg.dat, up.scan.name="swedishgenomes", win.size=10000)

scan.nonscan.af.plot.output(n.comp.af.boxplots=n.comp.af.boxplots, s.comp.af.boxplots=s.comp.af.boxplots, up.scan.name="swedishgenomes", win.size=10000)

```



### 9. Hancock climate associations
```{r generate analysis hancock associations data}
# prep data
up.scan.file <- "./data/003.selection.scans/Hancock.tophits.txt"
ha.dat <- read.table(up.scan.file, stringsAsFactors=FALSE, header=TRUE)
ha.dat <- ha.dat[,2:3]
colnames(ha.dat) <- c("chr", "pos")
## input is 2 columns giving chr and pos

select.scan.plots(ss.dat=ha.dat, up.scan.name="hancockscan", win.size=10000)
```

### 10. Christian Huber N/S sweeps
```{r generate analysis christian huber sweeps N/S}
# prep data
up.scan.file <- "./data/003.selection.scans/christian.intervals.csv"
ns.dat <- read.csv(up.scan.file, stringsAsFactors=FALSE)
ns.dat <- ns.dat[,1:2]
colnames(ns.dat) <- c("chr", "pos")
## input is 2 columns giving chr and pos

select.scan.plots(ss.dat=ns.dat, up.scan.name="hubersweep", win.size=10000)
```

### 11. Horton sweep signals


### 12. more datasets!!!

