---
title: "F6.figure6"
author: "Daniele Filiault"
date: "7/23/2021"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(glue)
library(ggplot2)
library(viridis)
library(RColorBrewer)
library(grid)
library(ggpubr)
library(ggh4x)
library(data.table)
library(gtable)
library(ggplotify)

```

## Introduction

Generates figure 6 of manuscript. 
Results of rotations done in script(s) 53 to do genome rotation to get empirical distributions of Kologorov-Smirnov test values obtained in script 52 testing selection scan genome subsets vs entire genome distributions of AF, AFD, home beta.  See also script 54 for initial analysis

## 01. load data

```{r load data}
scans <- c("n.scan.rot", "s.scan.rot", "n.gea.rot", "s.gea.rot","n.fit.rot", "s.fit.rot")
vars <- c("n.sweep.rot", "s.sweep.rot", "n.gea.rot", "s.gea.rot","n.fit.rot", "s.fit.rot")
all.dat <- as.list(1:length(scans))
for(up in 1:length(scans)){
  #print(up)
  up.scan <- scans[up]
  #print(up.scan)
  up.var <- vars[up]
  #print(up.var)
  up.files <- glue('./data/53.data/{up.scan}.{1:25}.Rdat')
  for(up.f in up.files){
    load(up.f)
    up.dat <- do.call(rbind, get(up.var))
    if(exists("out.dat")){out.dat <- rbind(out.dat, up.dat)
    }else{out.dat <- up.dat}
  }
  out.dat <- unique(out.dat)
  out.dat$scan.type <- up.scan
  all.dat[[up]] <- out.dat
  rm(out.dat)
}
all.dat <- do.call(rbind, all.dat)

all.dat$scan.type <-  substring(all.dat$scan.type,3)
all.dat$exp.site <- do.call(rbind,strsplit(all.dat$exp, "_"))[,1]
all.dat$year <- do.call(rbind,strsplit(all.dat$exp, "_"))[,2]

```

## 02. get number of rotations with larger KS than observed for all exp/year by scan combinations

```{r get nrotations smaller than observed}
expnames <- unique(all.dat$exp)
scannames <- unique(all.dat$scan)

scan.pvals <- as.list(1:length(expnames))
for(up in 1:length(expnames)){
  up.exp <- expnames[up]
  oe.dat <- sapply(1:length(scannames), function(up.s){
    up.scan <- scannames[up.s]
    up.dat <- all.dat[all.dat$scan==up.scan & all.dat$exp==up.exp,]
    n.af <- sum(up.dat$af.statistic < up.dat[up.dat$rotation=="observed","af.statistic"])
    n.afd <- sum(up.dat$afd.statistic < up.dat[up.dat$rotation=="observed","afd.statistic"])
    n.beta <- sum(up.dat$beta.statistic < up.dat[up.dat$rotation=="observed","beta.statistic"])
    out.dat <- unique(up.dat[,colnames(up.dat)%in%c("exp","scan","scan.type","exp.site", "year")])
    out.dat$n.af <- n.af
    out.dat$n.afd <- n.afd
    out.dat$n.beta <- n.beta
    return(out.dat)
      })
  oe.dat <- t(oe.dat)
  scan.pvals[[up]] <- oe.dat
}

scan.pvals <- do.call(rbind, scan.pvals)
scan.pvals <- apply(scan.pvals, 2, as.character)
scan.pvals <- as.data.frame(scan.pvals, stringsAsFactors=FALSE)
scan.pvals$scan.type <- gsub(".rot", "", scan.pvals$scan.type)
scan.pvals$n.af <- as.numeric(scan.pvals$n.af)
scan.pvals$n.afd <- as.numeric(scan.pvals$n.afd)
scan.pvals$n.beta <- as.numeric(scan.pvals$n.beta)

## get proportions of rotations < observed
n.rotations <- 1000
scan.pvals$p.af <- 1-(scan.pvals$n.af/n.rotations)
scan.pvals$p.afd <- 1-(scan.pvals$n.afd/n.rotations)
scan.pvals$p.beta <- 1-(scan.pvals$n.beta/n.rotations)

## set experiment names and levels
scan.pvals$exp.site <- gsub("ADA", "NA", scan.pvals$exp.site)
scan.pvals$exp.site <- gsub("RAM", "NM", scan.pvals$exp.site)
scan.pvals$exp.site <- gsub("RAT", "SR", scan.pvals$exp.site)
scan.pvals$exp.site <- gsub("ULL", "SU", scan.pvals$exp.site)
scan.pvals$exp.site <- factor(scan.pvals$exp.site, levels=c("NA", "NM", "SR", "SU"))
scan.pvals <- cbind(scan.pvals, geo = ifelse(scan.pvals$exp.site%in%c("NA","NM"), "North Experiments","South Experiments"))
scan.pvals$geo <- factor(scan.pvals$geo, levels=c("North Experiments", "South Experiments"))

```

## 03. make betas pval heatmap

```{r data prep heapmap beta pvals}
scan.pvals$sig.beta <- sapply(scan.pvals$p.beta, function(x){
  if(x<=0.001){xs <- "0.001"
  } else if(x<=0.01){xs <- "0.01"
  } else if(x<=0.050){xs <- "0.05"
  } else(xs <- "ns")
})

scan.pvals$scan.type <- gsub("gea", "genome-\nenvironment\nassociation", scan.pvals$scan.type)
scan.pvals$scan.type <- gsub("fit", "field\nfitness", scan.pvals$scan.type)
scan.pvals$scan.type <- gsub("scan", "selection\nscan", scan.pvals$scan.type)

```

```{r plot heatmap beta pvals}
f.sep <- 0.5
p.colors <- c(viridis(3, begin=0.55, direction=-1), "grey95")

beta.heatmap <- ggplot(scan.pvals, aes(1,1)) +
  geom_tile(aes(fill = sig.beta)) +
  facet_nested(scan.type+scan~exp.site + year,drop=T,space="free",scales="free") +
  geom_text(aes(label=sprintf("%0.3f", round(p.beta, digits = 3)))) +
  scale_fill_manual(values=p.colors) +
  theme_bw() +
  theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank(), axis.ticks.y=element_blank(), strip.text.y.right = element_text(angle = 0), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.spacing.x = unit(c(0, f.sep, 0 ,f.sep, 0, f.sep, 0) , 'lines'), panel.spacing.y = unit(c( 0,0,0,0,f.sep,0,0,f.sep,0,0,0,0,0,0) , 'lines'), strip.background = element_rect(colour="black", fill="grey75")) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0)) + 
  labs(fill = "empirical\npvalue")
```

## 04. make AF and AFD pval heatmap

```{r reformatting for af and afd}
allele.pvals <- scan.pvals
allele.pvals <- cbind(allele.pvals, geo.l = ifelse(allele.pvals$exp.site%in%c("NA","NM"), "North","South"))
allele.pvals <- allele.pvals[,c(2:3,9:10,14)]
ap.l <- melt(data=allele.pvals, id.vars=c("scan", "scan.type","geo.l"), variable.name="al",value.name="pval")

ap.l$sig.pval <- sapply(ap.l$pval, function(x){
  if(x<=0.001){xs <- "0.001"
  } else if(x<=0.01){xs <- "0.01"
  } else if(x<=0.050){xs <- "0.05"
  } else(xs <- "ns")
})

ap.l$al <- gsub("p.afd", "allele frequency difference", ap.l$al)
ap.l$al <- gsub("p.af", "home allele frequency", ap.l$al)
ap.l$al <- factor(ap.l$al, levels=c("home allele frequency", "allele frequency difference"))

ap.t$scan.type <- gsub("\n","", ap.t$scan.type)
```

```{r heapmap allele pvals}

allele.heatmap <- ggplot(ap.t, aes(1,1)) +
  geom_tile(aes(fill = sig.pval)) +
  facet_nested(scan.type+scan~al + geo.l) +
  geom_text(aes(label=sprintf("%0.3f", round(pval, digits = 3)))) +
  #geom_text(aes(label = pval)) +
  scale_fill_manual(values=p.colors) +
  theme_bw()  +
  theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank(), axis.ticks.y=element_blank(), strip.text.y.right = element_text(margin = margin(0,0,0,0, "lines")), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.spacing.x = unit(c(0, f.sep, 0) , 'lines'), panel.spacing.y = unit(c( 0,0,0,0,f.sep,0,0,f.sep,0,0,0,0,0,0) , 'lines'), legend.position = "none", strip.background = element_rect(colour="black", fill="grey75")) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0))

# remove right strip
plot.tab <- ggplotGrob(allele.heatmap)
r.names <- plot.tab$layout$name[grep("strip-r",plot.tab$layout$name)]

gtable_filter_remove <- function (x, name, trim = TRUE){
  matches <- !(x$layout$name %in% name)
  x$layout <- x$layout[matches, , drop = FALSE]
  x$grobs <- x$grobs[matches]
  if (trim) 
    x <- gtable_trim(x)
  x
}

allele.heatmap.grob <- gtable_filter_remove(plot.tab, r.names, trim=T)

ah.ggplot <-  ggplotify::as.ggplot(allele.heatmap.grob)

```

## 06. combine plots

```{r combine panels}

fig6 <- ggarrange(NULL, ah.ggplot, beta.heatmap, ncol=3, widths = c(0.08, 0.8, 2.1), labels=c("", "A", "B" ), hjust=c(1))

ggsave(fig6, file="./manuscript.figures/Figure6.jpg", width=14, height=8)
```













## 07. old plot stuff below here

```{r sandbox}

ap.t <- ap.l
ap.t$scan.type <- gsub("\n","", ap.t$scan.type)

test.heatmap <- ggplot(ap.t, aes(1,1)) +
  geom_tile(aes(fill = sig.pval)) +
  facet_nested(scan.type+scan~al + geo.l) +
  geom_text(aes(label=sprintf("%0.3f", round(pval, digits = 2)))) +
  #geom_text(aes(label = pval)) +
  scale_fill_manual(values=p.colors) +
  theme_bw()  +
  theme(axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank(), axis.title.y=element_blank(), axis.text.y=element_blank(), axis.ticks.y=element_blank(), strip.text.y.right = element_text(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.spacing.x = unit(c(0, f.sep, 0) , 'lines'), panel.spacing.y = unit(c( 0,0,0,0,f.sep,0,0,f.sep,0,0,0,0,0,0) , 'lines'), strip.text.y.right = element_blank(), legend.position = "none") +
  scale_x_continuous(expand = c(0, 0)) +
  scale_y_continuous(expand = c(0, 0))






plot.tab <- ggplotGrob(test.heatmap)
r.names <- plot.tab$layout$name[grep("strip-r",plot.tab$layout$name)]

gtable_filter_remove <- function (x, name, trim = TRUE){
  matches <- !(x$layout$name %in% name)
  x$layout <- x$layout[matches, , drop = FALSE]
  x$grobs <- x$grobs[matches]
  if (trim) 
    x <- gtable_trim(x)
  x
}

allele.heatmap.grob <- gtable_filter_remove(plot.tab, r.names, trim=T)

p = ggplotify::as.ggplot(allele.heatmap.grob)

fig6 <- ggarrange(NULL, p, beta.heatmap, ncol=3, widths = c(0.08, 0.8, 2.1), labels=c("", "A", "B" ), hjust=c(1))

ggsave(fig6, file="./manuscript.figures/Figure6.jpg", width=14, height=8)


```


## 03. plot these "empirical pvalues" by experiment, one plot per scan type for clarity

```{r generate AF/AFD and beta pval plots}
types <- unique(scan.pvals$scan.type)


type.af.plots <- as.list(1:3)
type.afd.plots <- as.list(1:3)

types.af <- c("A  selection scan", "B  genotype environment association", "C  field fitness")
pals <- c("Set2","Accent","Set1")
types.afd <- c("D", "E", "F")

for(up in 1:length(types)){
  up.t <- types[up]
  up.title.af <- types.af[up]
  up.title.afd <- types.afd[up]
  up.pal <- pals[up] 
  
  af.pplot <- ggplot(data=scan.pvals[scan.pvals$scan.type==up.t,], aes(x=(p.beta), y=(p.af), color=scan)) +
    geom_point() +
    #facet_grid(year~exp.site) +
    facet_nested(year~geo+exp.site) +
    scale_color_brewer(palette=up.pal, name="scan name") +
    #scale_colour_viridis_d(name="scan name") +
    xlab("proportion of rotations with KS less than observed (beta)") +
    ylab("proportion of rotations with \nKS less than observed \n(allele frequency)") +
    theme_bw() +
    ggtitle(up.title.af) +
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    xlim(0,1) +
    ylim(0,1)
  type.af.plots[[up]] <- af.pplot
  
  afd.pplot <- ggplot(data=scan.pvals[scan.pvals$scan.type==up.t,], aes(x=p.beta, y=p.afd, color=scan)) +
    geom_point() +
    #facet_grid(year~exp.site) +
    facet_nested(year~geo+exp.site) +
    scale_color_brewer(palette=up.pal, name="scan name") +
    xlab("proportion of rotations with KS less than observed (beta)") +
    ylab("proportion of rotations with \nKS less than observed \n(allele frequency difference)") +
    theme_bw() +
    ggtitle(up.title.afd)+
    theme(axis.text.x = element_text(angle = 45, hjust=1)) +
    xlim(0,1) +
    ylim(0,1)
  type.afd.plots[[up]] <- afd.pplot
  }

```

## 04. generate single figures from figures in 03
```{r consolidate AF figures}

af.figure <- ggarrange(type.af.plots[[1]] + rremove("ylab") + rremove("xlab"), type.af.plots[[2]] + rremove("ylab") + rremove("xlab"), type.af.plots[[3]] + rremove("ylab") + rremove("xlab"), # remove axis labels from plots
                    labels = NULL,
                    ncol = 1, nrow = 3,
                    common.legend = FALSE,
                    align = "hv", 
                    legend="none",
                    font.label = list(size = 10, color = "black", face = "bold", family = NULL, position = "top"))

af.figure <- annotate_figure(af.figure, left = textGrob("proportion of rotations with KS less than observed (allele frequency)", rot = 90, vjust = 1, gp = gpar(cex = 1)))

```

```{r consolidate AFD figures}

afd.figure <- ggarrange(type.afd.plots[[1]] + rremove("ylab") + rremove("xlab"), type.afd.plots[[2]] + rremove("ylab") + rremove("xlab"), type.afd.plots[[3]] + rremove("ylab") + rremove("xlab"), # remove axis labels from plots
                    labels = NULL,
                    ncol = 1, nrow = 3,
                    common.legend = FALSE,
                    align = "hv", 
                    legend="none",
                    font.label = list(size = 10, color = "black", face = "bold", family = NULL, position = "top"))

afd.figure <- annotate_figure(afd.figure, left = textGrob("proportion of rotations with KS less than observed (allele frequency difference)", rot = 90, vjust = 1, gp = gpar(cex = 1)))

```

```{r consolidate legends}
legend_scan <- get_legend(type.af.plots[[1]])
legend_gea <- get_legend(type.af.plots[[2]])
legend_fit <- get_legend(type.af.plots[[3]])

all.legends <- plot_grid(legend_scan, legend_gea, legend_fit, ncol=1, align = "v")
```

## 05. assemble final figure from these pieces

```{r assembled figure}
data.fig <- ggarrange(af.figure, NULL, afd.figure, ncol=3, common.legend=FALSE, widths=c(1, 0.07, 1))
data.fig <- annotate_figure(data.fig,bottom = textGrob("proportion of rotations with KS less than observed (home beta)", gp = gpar(cex = 1)))

assembled.fig <- ggarrange(data.fig, NULL, all.legends, ncol=3, common.legend=FALSE, widths = c(6,0.08, 0.9))

ggsave(assembled.fig, file="./manuscript.figures/Figure6.jpg", width=12, height=9)
```